---
title: "RFLOMICS-input-data"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{RFLOMICS-input-data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  warning = FALSE,
  message = FALSE,
  comment = "#>"
)

```

```{r, echo=FALSE}
library(RFLOMICS)
```

RFLOMICS treats three types of omics (RNAseq, proteomics and metabolomics) and several datasets per omics issued from similar experimental design. Each dataset must respect completeness constraints (see ?).

Below each input file is detailed :

## Experimental design file
<!-- ce fichier contient le plan experimental (conditions) plus des metadata par echantillons -->

This file provides the experimental design table. The first column indicates the sample names. Next columns indicate experimental conditions (called biological factors or replicate factors) and metadata (optional). Each row describes a sample by specifying the level for each experimental condition.
This file is edited by the user and must contain a header with sensible column names that correspond respectively to the factor names and metadata names.

It is advised that each factor's modality starts with a letter (for example, for a factor called Month, it is better to write M1 rather than 1).

```{r ImportDesign, output="asis"}
# ExpDesign.file <- file.path(path=path.package("RFLOMICS"), 
#                             file="/ExamplesFiles/ecoseed/condition.txt")
ExpDesign.file <- file.path(paste0(system.file(package = "RFLOMICS"), "/ExamplesFiles/ecoseed/condition.txt"))
ExpDesign      <- read.table(file = ExpDesign.file, header = TRUE, row.names = 1, sep = "\t")

DT::datatable(ExpDesign)
```

## Omic data file

For each omic : 
The matrix-like file with abundance of omic data is needed. The first column indicates feature names (genes, proteins, or metabolites), header indicates sample names.

### **RNAseq data**

<!-- les comptage de reads sont obtenu aprÃ¨s une analyse bioinfo (mapping sur le genome de ref et comptage des reads qui couvrent chaque gene) -->

For RNAseq data the values correspond to raw read counts per genes.

```{r RNAseq, output="asis"}

# geneCount.file <- file.path(path = path.package("RFLOMICS"), 
#                             file ="/ExamplesFiles/ecoseed/transcriptome_ecoseed.txt")
geneCount.file <- file.path(paste0(system.file(package = "RFLOMICS"),
                                   "/ExamplesFiles/ecoseed/transcriptome_ecoseed.txt"))
geneCount      <- read.table(file = geneCount.file, header = TRUE, row.names = 1, sep = "\t")

DT::datatable(geneCount[1:10, 1:5])
```


### **Proteomic data**

For proteomic data the values correspond to intensity of proteins.


```{r Proteomics, output="asis"}
# protAbundance.file <- file.path(path = path.package("RFLOMICS"), 
#                                 file ="/ExamplesFiles/ecoseed/proteome_ecoseed.txt")
protAbundance.file <- file.path(paste0(system.file(package = "RFLOMICS"),
                                   "/ExamplesFiles/ecoseed/proteome_ecoseed.txt"))
protAbundance <- read.table(file = protAbundance.file, header = TRUE, row.names = 1, sep = "\t")

DT::datatable(protAbundance[1:10, 1:5])
```


### **Metabolomic data**

For metabolomic data the values correspond to intensity of metabolites.

```{r Metabolomics, output="asis"}
# metaAbundance.file <- file.path(path = path.package("RFLOMICS"), 
#                                 file ="/ExamplesFiles/ecoseed/metabolome_ecoseed.txt")
metaAbundance.file <- file.path(paste0(system.file(package = "RFLOMICS"),
                                   "/ExamplesFiles/ecoseed/metabolome_ecoseed.txt"))
metaAbundance <- read.table(file = metaAbundance.file, header = TRUE, row.names = 1, sep = "\t")

DT::datatable(metaAbundance[1:10, 1:5])
```

## Annotation of features (optional)

This file contains annotation about biological fonctions of genes/proteins/metabolites, or their implication in biological pathways. This annotation is needed to compute Over Representation Analysis (ORA). This file must contain at least 2 columns, names of features same as used in abundance matrix and term identifiants (ex. GO term accession : GO:0034599). It is possible to add 2 more informations, names of terms (ex. cellular response to oxidative stress) and domain of annotation in supplementary colomns (ex. biological_process).

```{r, output="asis"}

# annot.genes.file <- file.path(path=path.package("RFLOMICS"), file="/ExamplesFiles/ecoseed/AT_GOterm_EnsemblPlants.txt")
annot.genes.file <- file.path(paste0(system.file(package = "RFLOMICS"),
                                   "/ExamplesFiles/ecoseed/AT_GOterm_EnsemblPlants.txt"))
annot.genes <- read.table(file = annot.genes.file, header = TRUE, sep="\t", )
colnames(annot.genes) <- c("geneID", "Term", "Name", "Domain")

DT::datatable(annot.genes[1:10,])
```

