---
title: "RFLOMICS"
output: 
  BiocStyle::html_document:
        toc: true
        toc_depth : 2
        number_sections: true
package: RFLOMICS
vignette: >
  %\VignetteIndexEntry{RFLOMICS}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
 
 

abstract: > 
   This document will go through a multi-omics data analysis with the RFLOMICS package. 
    It will be shown how to set up the statistical framework of the designed protocol: ie translate biological condition comparisons into contrasts, perform quality controls and statistical analysis of each omics dataset: ie. find differentially expressed genes/proteins/metabolites (DEF) per contrasts, identify clusters of co-expressed genes and perform a gene set enrichment analysis from lists of DEF or from clusters of co-expressed genes/proteins/metabolites. 
   
---


```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Run RFLOMICS

```{bash, eval=FALSE}
library(RFLOMICS)
RFLOMICS::runRFLOMICS()
```

On the left side of the interface, the different tab menus appears as the analysis progresses to guide the user.

# Load Data

We start by setting a project name (it is mandatory). Then we load the experimental design and the different omics datasets.

## Dataset example description:

The data used for this vignette have been provided by Pr. Loic Rajjou and Gwendal Cueff. It is included in the xxxx directory of the package. Briefly, A. thaliana's transcriptome, metabolome and proteome have been obtained in the context of the study of seed germination and vigor. 
In particular, the authors were interested in deciphering key entities involved in response to environmental stresses (on the mother plant): influences of temperature (high, medium and low) and imbibition stage (Dry: DI, early imbibition: EI and late imbibition: LI). See(https://www.uibk.ac.at/botany/ecoseed/home/) 



## Load Experimental Design

The experimental design file contains information about the readiness status of each sample (see link : vignette input ).
As soon as this file is loaded, we can define the type (Biological factor : Bio; batch factor : batch; metadata information : meta) and reference level of each design factor.

* *It is required to have minimum 1 biological factor (maximum 3 biological factors) and 1 batch factor. The order of the columns does not matter.*

* *It possible to exclude samples from design. These samples will be excluded from the analysis even if they are present in the omics dataset matrix.*

## Load Omics data

 For this tutorial we provide 3 datasets: RNAseq data (read counts), proteomics data (abundance of proteins), and metabolomics data (intensity of metabolites) (see link : vignette input). We must specify for each dataset its type of omics. We can give a suitable name to our dataset. By Default, they will be called respectively (set1, set2, set3).

As soon as the data are loaded the **RFlomics MultiAssayExperiment** object is created (see link : data management). The upset representation provides details about sample intersection of different datasets.  

![ ](Images/01_loadData.png){width=80%}

## Completeness check

 The design must be complete and as possible balanced. A **Complete design (mandatory)** mean that all possible combinations of biological factors' levels (called groups) are presents. A **Balanced design (recommended)** has an equal number of observations (replicates) for each group.

Thess constraints are checked and represented as graph (each square represent 1 group and each color level indicate the number of observations). 

```{r, fig.show="hold", out.width="50%", fig.align = "default", echo=FALSE}
knitr::include_graphics(c("Images/00_completeDesign.png", "Images/00_nonBalancedDesign.png"))
```

# Set the statistical framework

 The interface provides models written from the simplest one to the most complete one. Only the second order interaction terms between the biological factors appear. Batch factor never appears in interaction terms. In this case the complete model is chosen.

All 'a priori' hypothesis (contrast) calculated from the chosen model are displayed (see link vignette contrast). The user must select those that correspond to the biological issues.

In our case we are studying the effect of temperature on seed germination (transition from the state of dormancy to the state of germination vigor). For that we select 3 average contrasts (see image):

This statistical framework will be applied on all loaded datasets.  

![ ](Images/02_setModel.png){width=80%}

# Omics data analysis

Once the contrasts are chosen, 3 Item menus appear in the side bar menu that correspond to the 3 loaded omics types.

For each dataset, different steps of data analysis are proposed as a tab panel. These analysis steps must be performed sequentially. It is possible to switch between datasets.
However, you cannot perform a new task until the previous one is completed. A progress bar is displayed indicating the progress status.

## Data Exploratory

In this tab panel we access a data quality control screen and we perform the appropriate data filtering and processing according to the type of omics.

By default, features with 0 count in all conditions are removed from the data.

Specifically for each data we have the possibility to exclude samples or to select a subset of samples to analyze, provided the completeness condition is met. 

The effect of data filtering or sample removing step can be explored thanks to many QC graphs. For each QC, a graph is built on each raw data and filtered data to allow the comparison.


###  RNAseq data

For RNAseq data, low expressed genes are filtered so that only genes whose cpm expression is greater than **"CPM cutoff"** in x samples are kept. The value of x is given by **"Filtering strategy"** (NbOfsample_over_cpm <= NbConditions). By default the data are filtered based on NbCondition strategy with CPM cutoff equal to 5. In this example, there are 9 conditions. So, the number of genes with a CPM less than 5 in at least 9 samples are removed.

```{r}
# geneToKeep <- counts[rowSums(edgeR::cpm(counts) >= CPM_Cutoff) >= x, ]
```

To correct for differences in sequencing depths (library size), RFlomics proposes one method for normalization, **"TMM method"** from edgeR package.

```{r}
# dge <- edgeR::DGEList(counts=counts, group=groups)
# dge <- edgeR::calcNormFactors(dge, method="TMM")
```

For RNAseq data the QC graphs are:

- The library size distribution which
- Data Summary 
- Distribution
- Distribution density
- Principal component analysis

![ ](Images/03_DataProcess.png){width=80%}

## Differential expression analysis

In this tab panel, we can run the differential expression analysis. The analysis is run by contrast.  
For RNAseq data, the **glmFit()** function from the **edgeR** package is used whereas for proteomic and metabolomic data, the **lmFit()** function from **limma** packages is used. The ability to run analysis on a cluster is offered by sliding the cluster button (see RFLOMICS-configure-cluster-access vignette)


Results will appear in two components: a scrolling menu will give the results of the analysis by contrat and, 
at the bottom of the panel, the intersecting sets of the lists of DEG is given thanks to the **UpSetR()** function.  

Each results has to be validated:

- You **have to** to check the Pvalue distribution for each contrast and validate it (**ok checkbox**). 
  + If the distribution is uniform: it is ok. A few genes are DE and the FDR will find them.
  + If the distribution as a peak near 0 and then is uniform: it is also ok. The higher the peak, the higher
the number of DE genes. FDR correction can be applied.
  + If the distribution has 2 peaks one near 0 and one near 1 or just one peak near 1, it is not a good news! 
  You have to understand what's happened.

- You can filter differentially expressed genes/proteins/metabolites either by fold change (**|FC|**) or 
by False Discovery Rate (BH) cut-off. Graphs will be automatically updated.

- You have to **validate** to save the cut-off thresholds but also the contrasts' result selection  and pass
to the co-expression analysis or to the annotation enrichment step.
 
![ ](Images/04_diffAnal.png){width=80%}
 
## Coexpression analysis

* In this tab panel, choice is given to merge (**union**) or intersect  (**intersection**) the lists of differentially expressed 
genes/proteins/metabolites associated to contrasts.

* Co-expression analysis is done using to the **coseq** package with a set of optimal parameters.

  - For **RNAseq**, data transformation method is set to the **arcsin** function and normalization method to **TMM**. No
scaling is done for these data. **gaussian mixture model** is used to decipher the different profiles of gene expression with
a set of parameters to estimate fixed to (Gaussian_pK_Lk_Ck).
  - For **proteomics/metabolomics** data, scaling is done on proteins/metabolites. Another sets of parameters estimation is also proposed for the modeling of gene expression profile: (Gaussian_pK_Lk_Bk). It has to be used, if the other one doesn't fit.


The number of technical replicates to perform (**iter**) for each **K** which is the number of expressions' profile into which
the DEG have to be cut, can be first set to 2 to precise the range of K and then put to a minimum of 20 replicates per K.

The ability to run the clustering analysis on a cluster is offered by sliding the Cluster button (see Rflomics_configure-cluster-access).


* Results description (see [CoSeq](https://www.bioconductor.org/packages/release/bioc/vignettes/coseq/inst/doc/coseq.html))

  - Results can be explored with several plots proposed by **coseq**. The ICL graph has been slightly modified to show all the range of the ICL values for a given K. 
   
  - A table of jobs summary is also given. It groups error messages by K.  

![ ](Images/05_coExpAnal.png){width=80%}

## Annotation and Enrichment analysis using clusterProfilR

 We select the lists of DE genes or clusters to annotate. All available lists are selected by default. 

 The enrichment analysis is performed using the [clusterProfiler R-package](https://www.sciencedirect.com/science/article/pii/S2666675821000667). For more information on the methods and the package, please see [The article](https://www.sciencedirect.com/science/article/pii/S2666675821000667), [Bioconductor Package](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html) and the [clusterProfiler Vignette](https://yulab-smu.top/biomedical-knowledge-mining-book/). 

You can chose between several domains, depending on several parameters: 

* custom: preferred option. It requires an annotation file (tabulated, at least two columns, one for the name of the gene, the other for the name or id of the annotation term). A second panel will allow the user to indicate which columns indicates what. You can have as many domains/ontology as you want, you just need to indicate the columns name that differentiate them in the right area;
* GO (or GO:BP, GO:CC, GO:MF if only one GO domain is of interest). The user will have to chose a database library ([org.*.db](https://www.bioconductor.org/packages/release/BiocViews.html#___OrgDb)) installed on the libPath and the keytype of the entities (identifier to make the connection with the annotation). 
* KEGG: In the setting panel, enter the three letters to identify the organism and the keytype of the entities. It requires to have an online connection. 

Once everything is set and the analysis is conducted, several panels are displayed to show the results. 

![Results of the analysis for multiple lists](Images/06_ORAAnal.png){width=80%}

## Dataset analysis summary 

 This panel summarizes the effects of the filtering steps on the selected datasets and the results of the differential analyses. The first row shows how many samples and entities are left in each table. The integration will be performed on these datasets. 
The second row shows the number of DE entities for each selected contrast and for each table, with the distinction between up-regulated and down-regulated entities.  



# Integration

 Once at least two tables have passed the pre-processing step, the Integration tab will appear in the side bar menu. Two sub menus are available. For each of them, you will have to run a preprocessing before accessing to the integration panel: select the datatables you want to integrate together and the filtering strategy for each data. A filtering is strongly advised when having large data or unbalanced number of features between them. 
 
 You can filter according to the DE entities found in the differential analysis, or based on the variation coefficient, taking a certain number of variables that have the largest coefficient. For this example, we chose to keep only 500 features from the RNAseq data, based on the variation coefficient.
 
 RnaSeq data (counts transcriptomics) will be transformed before the integration. This is done with the **voom** transformation from the **limma** R package.   

 Batch effects are removed before applying the function, using **removebatcheffect** from **limma**.

![Integration data preparation](Images/07_prepData2.png){width=80%}

## MOFA

 This integration analysis is performed using the **MOFA2** R-package. It is an unsupervised method that searches for latent factors to decompose the variance of the data, similarly to a PCA, but applied to a multi-dataset situation. The interpretation of the results is also very similar to those of a PCA: the most interesting factors are the one explaining more variance, whether it's on several dataset or just one table. It also gives weights to the features of each table, for each factor. This is interesting in case a factor can be **a posteriori** linked to a covariate (in the case displayed on the figure, the temperature seems linked to the first factor, the features used to build this factor are to be considered very relevant in the analysis).  

![MOFA results on the ecoseed datasets](Images/07_MOFA.png){width=80%}

 The user can chose to scale or not each tables, with the scale views argument, then set the number of factors to be computed and finally the maximum iterations for the MOFA run. We recommend that the user do not change the number of iterations.  Once everything is set, you can run the analysis by clicking on the run button. 
 
Several figure are provided by the interface once the analysis is done.

## MixOmics

 This integration analysis is performed using the **MixOmics** R-package. In RFLOMICS, we focused on the supervised analyses using the functions block.(s)plda (also called DIABLO) from the package. MixOmics does not work on unbalanced design. Only complete cases will be used for the analysis. The user can select all biological factors as response variables. The meta and  batch factors are not available.

Using Sparse Analysis will set the "s" in the function names. It this case, the tuning cases argument will be considered and passed to tune.splsda or tune.block.splsda: it determines the number of number of variables to be tested: mixOmics block.splsda requires the argument test.keepX, which is a vector of number of variables to keep (eg : c(10, 50, 100)). Models using these number of features are then adjusted and the best model, according to the response variable, is selected and given to the user. 

In RFLOMICS, we chose to only ask for a number of cases. This number of cases is used to determine the keep.X, using a span of nfeatures/number of cases.

 Once everything is set, you can run the analysis by clicking on the run button. 
Several figures are provided by the interface once the analysis is done, for each selected response variable.  

![MixOmics Settings](Images/07_mixOmics.png){width=80%} 


