<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>RFLOMICS: Quick Guide to omics data analysis • RFLOMICS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="RFLOMICS: Quick Guide to omics data analysis">
<meta property="og:description" content="RFLOMICS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RFLOMICS</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/RFLOMICS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/RFLOMICS-input-data.html">RFLOMICS input data</a>
    </li>
    <li>
      <a href="../articles/RFLOMICS-contrast.html">RFLOMICS contrast</a>
    </li>
    <li>
      <a href="../articles/RFLOMICS-install.html">RFLOMICS install</a>
    </li>
    <li>
      <a href="../articles/RFLOMICS-data-management.html">RFLOMICS data management</a>
    </li>
    <li>
      <a href="../articles/RFLOMICS-benchmark-coseq.html">RFLOMICS benchmark coseq</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/RFLOMICS-command-lines.html">RFLOMICS command lines</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="RFLOMICS-command-lines_files/header-attrs-2.11/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>RFLOMICS: Quick Guide to omics data analysis</h1>
                        <h4 data-toc-skip class="author">Delphine Charif and Nadia Bessoltane</h4>
            
            <h4 data-toc-skip class="date">2021-10-06</h4>
      
      
      <div class="hidden name"><code>RFLOMICS-command-lines.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Here, we will walk through -omics data analysis with the RFLOMICS package. We will see how to set up the statistical framework of the designed protocol: ie translate biological condition comparisons into contrasts, perform quality controls and statistical analysis of the dataset: ie. find DEG per contrasts, identify cluster of co-expressed genes and perform a gene set enrichment analysis from lists of DEG or from clusters of co-expressed genes.</p>
    </div>
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RFLOMICS</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></code></pre></div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor" aria-hidden="true"></a>Introduction:</h1>
<p>RFLOMICS is based on proven statistical methods and pipeline for RNAseq data analysis (1). It is also based on functions allowing the automation of contrasts’s writting in R (2).</p>
<p>Data that will be used for this example have been provided by Loic Rajjou and Gwendal Cueff. They are included in the inst/ExampleFiles/ecoseed directory of the package. Briefly, A. thaliana’s transcriptomes have been obtained in the context of the study of seed germination and vigor. In particular, the author were interested in the influence of temperature (high, medium and low) and imbibition (Dry: DI, early imbibition: EI and late imbibition: LI) on gene’s expression.</p>
<ul>
<li><ol style="list-style-type: decimal">
<li><a href="http://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=pubmed&amp;id=32426025&amp;retmode=ref&amp;cmd=prlinks" class="external-link">Ilana Lambert, Christine Paysant-Le Roux, Stefano Colella, and Marie-Laure Martin-Magniette (2020) DiCoExpress: a tool to process multifactorial RNAseq experiments from quality controls to co-expression analysis through differential analysis based on contrasts inside GLM models. <em>Plant Methods</em> vol. 16, p. 68</a></li>
</ol></li>
<li><ol start="2" style="list-style-type: decimal">
<li>[Christine Paysant-Le Roux (2020), packageContrast_2020, unpublished]</li>
</ol></li>
<li><ol start="3" style="list-style-type: decimal">
<li>[Projet ecoseed (), ]</li>
</ol></li>
</ul>
</div>
<div id="input-data" class="section level1">
<h1 class="hasAnchor">
<a href="#input-data" class="anchor" aria-hidden="true"></a>Input data</h1>
<ul>
<li>
<strong>Set</strong> name project.</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">projectName</span> <span class="op">=</span> <span class="st">"ecossed"</span></code></pre></div>
<ul>
<li>
<strong>Load</strong> the experimental design file.</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ExpDesign</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span>file<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/find.package.html" class="external-link">path.package</a></span><span class="op">(</span><span class="st">"RFLOMICS"</span><span class="op">)</span>, <span class="st">"/ExamplesFiles/ecoseed/condition.txt"</span>,sep<span class="op">=</span><span class="st">""</span><span class="op">)</span>,
                          header <span class="op">=</span> <span class="cn">TRUE</span>,row.names <span class="op">=</span> <span class="fl">1</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></code></pre></div>
<ul>
<li>
<strong>Load</strong> the omics data files</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">geneCount</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span>file   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span>path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/find.package.html" class="external-link">path.package</a></span><span class="op">(</span><span class="st">"RFLOMICS"</span><span class="op">)</span>, file <span class="op">=</span><span class="st">"/ExamplesFiles/ecoseed/transcriptome2_ecoseed.txt"</span><span class="op">)</span>,
                        header <span class="op">=</span> <span class="cn">TRUE</span>, row.names <span class="op">=</span> <span class="fl">1</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span>

<span class="co"># protAbundance &lt;- read.table(file = file.path(path = path.package("RFLOMICS"), file ="/ExamplesFiles/ecoseed/proteome_ecoseed.txt"),</span>
<span class="co">#                             header = TRUE,row.names = 1, sep = "\t")</span>
<span class="co"># metaAbundance &lt;- read.table(file = file.path(path = path.package("RFLOMICS"), file ="/ExamplesFiles/ecoseed/metabolome_ecoseed.txt"),</span>
<span class="co">#                             header = TRUE,row.names = 1, sep = "\t")</span></code></pre></div>
<ul>
<li>
<strong>Load</strong> annotation file(s) for enrichment analysis.</li>
</ul>
<p>The column 1 =&gt; feature names (same used in abundance matrix) The column 2 =&gt; term IDs (ex. GO term accession : <a href="GO:0034599" class="uri">GO:0034599</a>) The column 3 =&gt; term names (ex. GO term name : cellular response to oxidative stress) The column 4 =&gt; source or type (ex. GO domain : biological_process)</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">annot.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span>file   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span>path<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/find.package.html" class="external-link">path.package</a></span><span class="op">(</span><span class="st">"RFLOMICS"</span><span class="op">)</span>, file<span class="op">=</span><span class="st">"/ExamplesFiles/ecoseed/AT_GOterm_EnsemblPlants.txt"</span><span class="op">)</span>, 
                          header <span class="op">=</span> <span class="cn">TRUE</span>, sep<span class="op">=</span><span class="st">"\t"</span>, <span class="op">)</span>
<span class="co">#&gt; Warning in scan(file = file, what = what, sep = sep, quote = quote, dec = dec, :</span>
<span class="co">#&gt; Fin de fichier (EOF) dans une chaîne de caractères entre guillements</span>
<span class="co">#&gt; Warning in scan(file = file, what = what, sep = sep, quote = quote, dec = dec, :</span>
<span class="co">#&gt; le nombre d'objets lus n'est pas un multiple du nombre de colonnes</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">annot.genes</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"geneID"</span>, <span class="st">"Term"</span>, <span class="st">"Name"</span>, <span class="st">"Domain"</span><span class="op">)</span></code></pre></div>
</div>
<div id="set-up-the-statistical-framework" class="section level1">
<h1 class="hasAnchor">
<a href="#set-up-the-statistical-framework" class="anchor" aria-hidden="true"></a>Set up the statistical framework</h1>
<ul>
<li>Define the type of factors (either ‘biological’=‘Bio’ or ‘batch’=‘batch’ effect factor) and the level of reference for each factor.</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Design.Factors.Name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ExpDesign</span><span class="op">)</span>

<span class="va">Design.Factors.Type</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"batch"</span>,<span class="st">"Bio"</span>,<span class="st">"Bio"</span><span class="op">)</span>

<span class="co"># Define the level of reference (factor modality) of each factor</span>
<span class="va">Design.Factors.Ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rep1"</span>,<span class="st">"Medium"</span>,<span class="st">"EI"</span><span class="op">)</span>

<span class="co"># remplacer les 2 derniers parametres par un dataframe? </span>
<span class="co"># Design.Factors &lt;- data.frame(Design.Factors.Name = c("Repeat", "temperature", "imbibition"),</span>
<span class="co">#                              Design.Factors.Type = c("batch","Bio","Bio"),</span>
<span class="co">#                              Design.Factors.Ref  = c("rep1","Medium","EI"))</span></code></pre></div>
<ul>
<li>Instantiate an object of class Design. This object will contain all information about statistical framework.</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Design &lt;- ExpDesign.constructor(ExpDesign = ExpDesign, Design.Factors = Design.Factors)</span>

<span class="va">Design</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExpDesign-Constructor.html">ExpDesign.constructor</a></span><span class="op">(</span>ExpDesign <span class="op">=</span> <span class="va">ExpDesign</span>, refList <span class="op">=</span> <span class="va">Design.Factors.Ref</span>, typeList <span class="op">=</span> <span class="va">Design.Factors.Type</span><span class="op">)</span></code></pre></div>
<ul>
<li>Check the <strong>completeness</strong> of the design. You <strong>must</strong> have all possible combinations of factor’s level. <strong>Balanced design</strong> is not required (presence of the same number of replicats for all possible combinations) but advised. You <strong>must</strong> also have at least one biological factor and 2 replicats.</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R">  <span class="co"># check completeness</span>
  <span class="va">completeCheckRes</span> <span class="op">&lt;-</span> <span class="fu">CheckExpDesignCompleteness</span><span class="op">(</span><span class="va">Design</span><span class="op">)</span>

  <span class="co"># plot</span>
  <span class="va">completeCheckRes</span><span class="op">$</span><span class="va">plot</span></code></pre></div>
<div class="figure">
<img src="RFLOMICS-command-lines_files/figure-html/CheckDesign-1.png" alt="Design checking's result" width="700"><p class="caption">
Design checking’s result
</p>
</div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R">  <span class="co"># error message if exist</span>
  <span class="va">completeCheckRes</span><span class="op">$</span><span class="va">error</span>
<span class="co">#&gt; NULL</span></code></pre></div>
<p>In this example, you can see that the design is complete but not balanced.</p>
<ul>
<li>
<strong>Choose</strong> the statistical model associated to the experimental design. Models are written from the most complete one, with two orders interaction terms between the biological factors, to the simplest one. Batch factor did not appears in interaction terms.</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># remplacer aussi ici</span>
<span class="co">#Design.formulae &lt;- GetModelFormulae(Design.Factors = Design.Factors)</span>

<span class="va">Design.formulae</span> <span class="op">&lt;-</span> <span class="fu">GetModelFormulae</span><span class="op">(</span>Factors.Name <span class="op">=</span> <span class="va">Design.Factors.Name</span>,
                                    Factors.Type <span class="op">=</span> <span class="va">Design.Factors.Type</span><span class="op">)</span>

 <span class="va">Design.formulae</span>
<span class="co">#&gt; $`~Repeat + temperature + imbibition + temperature:imbibition`</span>
<span class="co">#&gt; ~Repeat + temperature + imbibition + temperature:imbibition</span>
<span class="co">#&gt; &lt;environment: 0x7fc4b1b7add0&gt;</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $`~Repeat + temperature + imbibition`</span>
<span class="co">#&gt; ~Repeat + temperature + imbibition</span>
<span class="co">#&gt; &lt;environment: 0x7fc4b1b6a040&gt;</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $`~Repeat + temperature`</span>
<span class="co">#&gt; ~Repeat + temperature</span>
<span class="co">#&gt; &lt;environment: 0x7fc4b1ba4f20&gt;</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $`~Repeat + imbibition`</span>
<span class="co">#&gt; ~Repeat + imbibition</span>
<span class="co">#&gt; &lt;environment: 0x7fc4b1b8a318&gt;</span></code></pre></div>
<p>In this example, the complete model with the interaction term describes the experiment.</p>
<ul>
<li>
<p><strong>View</strong> all the ‘a priori’ hypothesis you thought for the experiment. They are divided into 3 categories: simple, averaged and interaction contrasts</p>
<ul>
<li>
<strong>simple</strong>: For a given biological factor, does a difference in gene expression between two factor’s modalities exists when one modality has been fixed for all other factors.</li>
<li>
<strong>averaged</strong>: Does the difference in gene expression between two factor’s modalities exists if we consider an average gene expression between the modalities for the other factor’s ?</li>
<li>
<strong>interaction</strong>: Does the difference in gene expression between two factor’s modalities is the same for the different modalities of other factors ?</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"> <span class="co"># Get all the possible contrasts from the complete model with interaction</span>
 <span class="va">Design</span> <span class="op">&lt;-</span> <span class="fu">getExpressionContrast</span><span class="op">(</span>object <span class="op">=</span> <span class="va">Design</span>, 
                                 model.formula <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">Design.formulae</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; treatment factors names:</span>
<span class="co">#&gt; [1] "temperature" "imbibition"</span>
 <span class="co"># Simple Contrasts</span>
 <span class="co"># knitr::kable( Design@Contrasts.List$simple[,-c(1:2)], caption = 'Simple Contrasts') </span>
  <span class="va">Design</span><span class="op">@</span><span class="va">Contrasts.List</span><span class="op">$</span><span class="va">simple</span><span class="op">[</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span><span class="op">]</span>
<span class="co">#&gt;                                                  contrastName   type</span>
<span class="co">#&gt;  1:    (temperatureLow - temperatureElevated) in imbibitionLI simple</span>
<span class="co">#&gt;  2: (temperatureMedium - temperatureElevated) in imbibitionLI simple</span>
<span class="co">#&gt;  3:      (temperatureMedium - temperatureLow) in imbibitionLI simple</span>
<span class="co">#&gt;  4:    (temperatureLow - temperatureElevated) in imbibitionEI simple</span>
<span class="co">#&gt;  5: (temperatureMedium - temperatureElevated) in imbibitionEI simple</span>
<span class="co">#&gt;  6:      (temperatureMedium - temperatureLow) in imbibitionEI simple</span>
<span class="co">#&gt;  7:    (temperatureLow - temperatureElevated) in imbibitionDS simple</span>
<span class="co">#&gt;  8: (temperatureMedium - temperatureElevated) in imbibitionDS simple</span>
<span class="co">#&gt;  9:      (temperatureMedium - temperatureLow) in imbibitionDS simple</span>
<span class="co">#&gt; 10:      (imbibitionEI - imbibitionLI) in temperatureElevated simple</span>
<span class="co">#&gt; 11:           (imbibitionEI - imbibitionLI) in temperatureLow simple</span>
<span class="co">#&gt; 12:        (imbibitionEI - imbibitionLI) in temperatureMedium simple</span>
<span class="co">#&gt; 13:      (imbibitionDS - imbibitionLI) in temperatureElevated simple</span>
<span class="co">#&gt; 14:           (imbibitionDS - imbibitionLI) in temperatureLow simple</span>
<span class="co">#&gt; 15:        (imbibitionDS - imbibitionLI) in temperatureMedium simple</span>
<span class="co">#&gt; 16:      (imbibitionDS - imbibitionEI) in temperatureElevated simple</span>
<span class="co">#&gt; 17:           (imbibitionDS - imbibitionEI) in temperatureLow simple</span>
<span class="co">#&gt; 18:        (imbibitionDS - imbibitionEI) in temperatureMedium simple</span>
<span class="co">#&gt;               fixFactor</span>
<span class="co">#&gt;  1:        imbibitionLI</span>
<span class="co">#&gt;  2:        imbibitionLI</span>
<span class="co">#&gt;  3:        imbibitionLI</span>
<span class="co">#&gt;  4:        imbibitionEI</span>
<span class="co">#&gt;  5:        imbibitionEI</span>
<span class="co">#&gt;  6:        imbibitionEI</span>
<span class="co">#&gt;  7:        imbibitionDS</span>
<span class="co">#&gt;  8:        imbibitionDS</span>
<span class="co">#&gt;  9:        imbibitionDS</span>
<span class="co">#&gt; 10: temperatureElevated</span>
<span class="co">#&gt; 11:      temperatureLow</span>
<span class="co">#&gt; 12:   temperatureMedium</span>
<span class="co">#&gt; 13: temperatureElevated</span>
<span class="co">#&gt; 14:      temperatureLow</span>
<span class="co">#&gt; 15:   temperatureMedium</span>
<span class="co">#&gt; 16: temperatureElevated</span>
<span class="co">#&gt; 17:      temperatureLow</span>
<span class="co">#&gt; 18:   temperatureMedium</span>
 <span class="co"># Averaged Contrasts</span>
 <span class="co"># knitr::kable( Design@Contrasts.List$averaged[,-c(1:2)], caption = 'Averaged Contrasts')</span>
 <span class="va">Design</span><span class="op">@</span><span class="va">Contrasts.List</span><span class="op">$</span><span class="va">averaged</span><span class="op">[</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span><span class="op">]</span>
<span class="co">#&gt;                                         contrastName type</span>
<span class="co">#&gt; 1:    (temperatureLow - temperatureElevated) in mean mean</span>
<span class="co">#&gt; 2: (temperatureMedium - temperatureElevated) in mean mean</span>
<span class="co">#&gt; 3:      (temperatureMedium - temperatureLow) in mean mean</span>
<span class="co">#&gt; 4:             (imbibitionEI - imbibitionLI) in mean mean</span>
<span class="co">#&gt; 5:             (imbibitionDS - imbibitionLI) in mean mean</span>
<span class="co">#&gt; 6:             (imbibitionDS - imbibitionEI) in mean mean</span>
<span class="co">#&gt;                                                      meanIn</span>
<span class="co">#&gt; 1:               imbibitionLI + imbibitionEI + imbibitionDS</span>
<span class="co">#&gt; 2:               imbibitionLI + imbibitionEI + imbibitionDS</span>
<span class="co">#&gt; 3:               imbibitionLI + imbibitionEI + imbibitionDS</span>
<span class="co">#&gt; 4: temperatureElevated + temperatureLow + temperatureMedium</span>
<span class="co">#&gt; 5: temperatureElevated + temperatureLow + temperatureMedium</span>
<span class="co">#&gt; 6: temperatureElevated + temperatureLow + temperatureMedium</span>
 <span class="co"># Interaction Contrasts</span>
 <span class="co"># knitr::kable(Design@Contrasts.List$interaction[,-c(1:2,4)], caption = 'Interaction Contrasts')</span>
 <span class="va">Design</span><span class="op">@</span><span class="va">Contrasts.List</span><span class="op">$</span><span class="va">interaction</span><span class="op">[</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="fl">4</span><span class="op">)</span><span class="op">]</span>
<span class="co">#&gt;             groupInteraction</span>
<span class="co">#&gt; 1: temperature vs imbibition</span>
<span class="co">#&gt; 2: temperature vs imbibition</span>
<span class="co">#&gt; 3: temperature vs imbibition</span>
<span class="co">#&gt; 4: temperature vs imbibition</span>
<span class="co">#&gt; 5: temperature vs imbibition</span>
<span class="co">#&gt; 6: temperature vs imbibition</span>
<span class="co">#&gt; 7: temperature vs imbibition</span>
<span class="co">#&gt; 8: temperature vs imbibition</span>
<span class="co">#&gt; 9: temperature vs imbibition</span>
<span class="co">#&gt;                                                                                                             contrastName</span>
<span class="co">#&gt; 1:       (temperatureLow - temperatureElevated) in imbibitionEI - (temperatureLow - temperatureElevated) in imbibitionLI</span>
<span class="co">#&gt; 2: (temperatureMedium - temperatureElevated) in imbibitionEI - (temperatureMedium - temperatureElevated) in imbibitionLI</span>
<span class="co">#&gt; 3:           (temperatureMedium - temperatureLow) in imbibitionEI - (temperatureMedium - temperatureLow) in imbibitionLI</span>
<span class="co">#&gt; 4:       (temperatureLow - temperatureElevated) in imbibitionDS - (temperatureLow - temperatureElevated) in imbibitionLI</span>
<span class="co">#&gt; 5: (temperatureMedium - temperatureElevated) in imbibitionDS - (temperatureMedium - temperatureElevated) in imbibitionLI</span>
<span class="co">#&gt; 6:           (temperatureMedium - temperatureLow) in imbibitionDS - (temperatureMedium - temperatureLow) in imbibitionLI</span>
<span class="co">#&gt; 7:       (temperatureLow - temperatureElevated) in imbibitionDS - (temperatureLow - temperatureElevated) in imbibitionEI</span>
<span class="co">#&gt; 8: (temperatureMedium - temperatureElevated) in imbibitionDS - (temperatureMedium - temperatureElevated) in imbibitionEI</span>
<span class="co">#&gt; 9:           (temperatureMedium - temperatureLow) in imbibitionDS - (temperatureMedium - temperatureLow) in imbibitionEI</span>
<span class="co">#&gt;           type</span>
<span class="co">#&gt; 1: interaction</span>
<span class="co">#&gt; 2: interaction</span>
<span class="co">#&gt; 3: interaction</span>
<span class="co">#&gt; 4: interaction</span>
<span class="co">#&gt; 5: interaction</span>
<span class="co">#&gt; 6: interaction</span>
<span class="co">#&gt; 7: interaction</span>
<span class="co">#&gt; 8: interaction</span>
<span class="co">#&gt; 9: interaction</span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"> <span class="co"># Select The 2 firsts</span>
 <span class="va">Design.contrastList</span> <span class="op">&lt;-</span> <span class="va">Design</span><span class="op">@</span><span class="va">Contrasts.List</span>
 <span class="co">#Design.contrastList$simple &lt;- Design.contrastList$simple[c(1:9)]$contrast</span>
 <span class="co">#Design.contrastList$averaged &lt;- Design.contrastList$averaged[1:3]</span>
 <span class="co">#Design.contrastList &lt;- Design.contrastList[-3]</span>

 <span class="co"># Set the coefficients</span>
 <span class="va">Design</span> <span class="op">&lt;-</span> <span class="fu">getContrastMatrix</span><span class="op">(</span>object <span class="op">=</span> <span class="va">Design</span>, contrastList <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">Design.contrastList</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="creating-a-flomics-multiassayexperiment-object" class="section level1">
<h1 class="hasAnchor">
<a href="#creating-a-flomics-multiassayexperiment-object" class="anchor" aria-hidden="true"></a>Creating a Flomics <strong>MultiAssayExperiment</strong> object</h1>
<ul>
<li><p>All input data (abundance, metadata, statistical framework), choice about tools, parameters, threshold, and analysis results are managed in (by) the FlomicsMultiAssayExp object, an instance of the class MultiAssayExperiment (MAE). See the next paragraph to have a description. This object is created by FlomicsMultiAssay.constructor method. This method takes in input : a list of lists (per dataset) of abundance matrices (and their metadata if it exist), and “Design” object.</p></li>
<li><p>Build a list of abundances :</p></li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ListofData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>
  <span class="st">"RNAseq1"</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span> <span class="st">"data"</span><span class="op">=</span><span class="va">geneCount</span>,      <span class="st">"meta"</span><span class="op">=</span><span class="cn">NULL</span>, <span class="st">"omicType"</span><span class="op">=</span><span class="st">"RNAseq"</span><span class="op">)</span><span class="op">)</span>
  <span class="co"># "proteomics1"=  list( "data"=protAbundance,  "meta"=NULL, "omicType"="proteomics"),</span>
  <span class="co"># "Metabolomics1"=list( "data"= metaAbundance, "meta"=NULL, "omicType"="Metabolomics"))</span></code></pre></div>
<ul>
<li>Instantiate an object of class MultiAssayExperiment.</li>
</ul>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"> <span class="va">FlomicsMultiAssay</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FlomicsMultiAssay.constructor.html">FlomicsMultiAssay.constructor</a></span><span class="op">(</span>inputs      <span class="op">=</span> <span class="va">ListofData</span>, 
                                                    Design      <span class="op">=</span> <span class="va">Design</span>, 
                                                    projectName <span class="op">=</span> <span class="va">projectName</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">FlomicsMultiAssay</span>
<span class="co">#&gt; A MultiAssayExperiment object of 1 listed</span>
<span class="co">#&gt;  experiment with a user-defined name and respective class.</span>
<span class="co">#&gt;  Containing an ExperimentList class object of length 1:</span>
<span class="co">#&gt;  [1] RNAseq1: SummarizedExperiment with 25014 rows and 27 columns</span>
<span class="co">#&gt; Functionality:</span>
<span class="co">#&gt;  experiments() - obtain the ExperimentList instance</span>
<span class="co">#&gt;  colData() - the primary/phenotype DataFrame</span>
<span class="co">#&gt;  sampleMap() - the sample coordination DataFrame</span>
<span class="co">#&gt;  `$`, `[`, `[[` - extract colData columns, subset, or experiment</span>
<span class="co">#&gt;  *Format() - convert into a long or wide DataFrame</span>
<span class="co">#&gt;  assays() - convert ExperimentList to a SimpleList of matrices</span>
<span class="co">#&gt;  exportClass() - save all data to files</span></code></pre></div>
</div>
<div id="statistical-analysis-of-rnaseq-data" class="section level1">
<h1 class="hasAnchor">
<a href="#statistical-analysis-of-rnaseq-data" class="anchor" aria-hidden="true"></a>Statistical analysis of RNAseq data:</h1>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data</span><span class="op">=</span><span class="st">"RNAseq1"</span>
<span class="va">RNAseq.raw.SE</span> <span class="op">&lt;-</span>  <span class="va">FlomicsMultiAssay</span><span class="op">@</span><span class="va">ExperimentList</span><span class="op">[[</span><span class="va">data</span><span class="op">]</span><span class="op">]</span></code></pre></div>
<!-- * dans le cas ou le nombre de samples dans la matrice ne correspond tous à fait aux samples decrits dans le design -->
<!-- il faut faire une fonction pour aider l'utilisateur de retirer des samples facilement -->
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">CheckCompleteness.dataset</span> <span class="op">&lt;-</span> <span class="fu">CheckExpDesignCompleteness</span><span class="op">(</span><span class="va">Design</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">RNAseq.raw.SE</span><span class="op">)</span><span class="op">)</span>

<span class="va">CheckCompleteness.dataset</span><span class="op">$</span><span class="va">plot</span></code></pre></div>
<p><img src="RFLOMICS-command-lines_files/figure-html/CheckExp-1.png" width="700"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">CheckCompleteness.dataset</span><span class="op">$</span><span class="va">error</span>
<span class="co">#&gt; NULL</span></code></pre></div>
<div id="filter-and-normalize" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-and-normalize" class="anchor" aria-hidden="true"></a>Filter and Normalize</h2>
<ul>
<li>By default, gene with 0 count are removed from the data.</li>
<li>The advice filtering method is to remove genes for which NbOfsample_over_cpm &lt;= NbConditions. For a CPM fixed at 5. In this example, there are 9 conditions. So, the number of gene with a CPM less than 5 in at least 9 samples are removed. Other strategies are available (see the help).</li>
</ul>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">RNAseq.SE</span> <span class="op">&lt;-</span> <span class="fu">FilterLowAbundance</span><span class="op">(</span><span class="va">RNAseq.raw.SE</span>, Filter_Strategy <span class="op">=</span> <span class="st">"NbConditions"</span>, CPM_Cutoff <span class="op">=</span> <span class="fl">5</span> <span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">RNAseq.SE</span><span class="op">@</span><span class="va">metadata</span><span class="op">)</span>
<span class="co">#&gt; [1] "omicType"         "Groups"           "rowSums.zero"     "FilteredFeatures"</span>
<span class="co">#&gt; [5] "FilteredOptions"  "FilteringOptions"</span></code></pre></div>
<p>We can obtain the number of genes which were filtred with this command:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">RNAseq.SE</span><span class="op">@</span><span class="va">metadata</span><span class="op">$</span><span class="va">FilteredFeatures</span><span class="op">)</span>
<span class="co">#&gt; [1] 11597</span></code></pre></div>
<p>11597 genes were filtered out.</p>
<ul>
<li>Run the TMM normalization method. Scaling factors are stored in the table below:</li>
</ul>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">RNAseq.SE</span> <span class="op">&lt;-</span> <span class="fu">RunNormalization</span><span class="op">(</span><span class="va">RNAseq.SE</span>, NormMethod <span class="op">=</span> <span class="st">"TMM"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">RNAseq.SE</span><span class="op">@</span><span class="va">metadata</span><span class="op">)</span>
<span class="co">#&gt; [1] "omicType"         "Groups"           "rowSums.zero"     "FilteredFeatures"</span>
<span class="co">#&gt; [5] "FilteredOptions"  "FilteringOptions" "Normalization"</span></code></pre></div>
</div>
<div id="exploratory" class="section level2">
<h2 class="hasAnchor">
<a href="#exploratory" class="anchor" aria-hidden="true"></a>Exploratory</h2>
<ul>
<li>Library size distribution of raw RNAseq data:</li>
</ul>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">Library_size_barplot.plot</span><span class="op">(</span><span class="va">RNAseq.raw.SE</span><span class="op">)</span>

<span class="fu">Library_size_barplot.plot</span><span class="op">(</span><span class="va">RNAseq.SE</span><span class="op">)</span>

<span class="va">CoefNorm</span> <span class="op">&lt;-</span> <span class="va">RNAseq.SE</span><span class="op">@</span><span class="va">metadata</span><span class="op">$</span><span class="va">Normalization</span><span class="op">$</span><span class="va">coefNorm</span>
<span class="va">CoefNorm</span><span class="op">$</span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">RNAseq.SE</span><span class="op">@</span><span class="va">metadata</span><span class="op">$</span><span class="va">Normalization</span><span class="op">$</span><span class="va">coefNorm</span><span class="op">)</span>
<span class="va">CoefNorm</span><span class="op">$</span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">CoefNorm</span><span class="op">$</span><span class="va">samples</span>, levels <span class="op">=</span> <span class="va">CoefNorm</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span>
<span class="va">pA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">CoefNorm</span>, <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/aes.html" class="external-link">aes</a></span><span class="op">(</span> x <span class="op">=</span> <span class="va">samples</span>, y <span class="op">=</span> <span class="va">lib.size</span> <span class="op">*</span> <span class="va">norm.factors</span>, fill <span class="op">=</span> <span class="va">group</span> <span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Library Size"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/theme.html" class="external-link">theme</a></span><span class="op">(</span> axis.text.x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, legend.position  <span class="op">=</span> <span class="st">"none"</span> <span class="op">)</span>
<span class="va">pA</span></code></pre></div>
<p><img src="RFLOMICS-command-lines_files/figure-html/Library%20size-1.png" width="700"><img src="RFLOMICS-command-lines_files/figure-html/Library%20size-2.png" width="700"><img src="RFLOMICS-command-lines_files/figure-html/Library%20size-3.png" width="700"></p>
<ul>
<li>Count distribution of raw RNAseq data:</li>
</ul>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Data_Distribution.plot(RNAseq.raw.SE, "density") </span>
<span class="co">#Data_Distribution.plot(RNAseq.raw.SE, "boxplot") </span>

<span class="fu">Data_Distribution_Density.plot</span><span class="op">(</span><span class="va">RNAseq.raw.SE</span><span class="op">)</span> 

<span class="fu">Data_Distribution_Density.plot</span><span class="op">(</span><span class="va">RNAseq.SE</span><span class="op">)</span> </code></pre></div>
<p><img src="RFLOMICS-command-lines_files/figure-html/CountDistr-1.png" width="700"><img src="RFLOMICS-command-lines_files/figure-html/CountDistr-2.png" width="700"></p>
</div>
<div id="technical-and-biological-variability-exploration-after-filtering-and-normalisation-step" class="section level2">
<h2 class="hasAnchor">
<a href="#technical-and-biological-variability-exploration-after-filtering-and-normalisation-step" class="anchor" aria-hidden="true"></a>Technical and Biological variability exploration after filtering and normalisation step</h2>
<ul>
<li>Raw data Exploratory of raw RNAseq data:</li>
</ul>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">RNAseq.raw.SE</span> <span class="op">&lt;-</span> <span class="fu">RunPCA</span><span class="op">(</span><span class="va">RNAseq.raw.SE</span><span class="op">)</span>
<span class="va">RNAseq.SE</span>     <span class="op">&lt;-</span> <span class="fu">RunPCA</span><span class="op">(</span><span class="va">RNAseq.SE</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># mvQCdesign(object = FlomicsMultiAssay, data = data , PCA = "raw", axis = 5, pngFile = NULL)</span>
<span class="co"># </span>
<span class="co"># mvQCdesign(object = FlomicsMultiAssay, data = paste0(data, ".filtred") , PCA = "norm", axis = 5, pngFile = NULL)</span></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">RFLOMICS</span><span class="fu">::</span><span class="fu">plotPCA</span><span class="op">(</span> <span class="va">RNAseq.raw.SE</span>, 
                   PCA <span class="op">=</span> <span class="st">"raw"</span>, PCs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, condition <span class="op">=</span> <span class="st">"groups"</span><span class="op">)</span>
<span class="fu">RFLOMICS</span><span class="fu">::</span><span class="fu">plotPCA</span><span class="op">(</span> <span class="va">RNAseq.SE</span>, 
                   PCA <span class="op">=</span> <span class="st">"norm"</span>, PCs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, condition <span class="op">=</span> <span class="st">"groups"</span><span class="op">)</span>


<span class="fu">RFLOMICS</span><span class="fu">::</span><span class="fu">plotPCA</span><span class="op">(</span> <span class="va">RNAseq.raw.SE</span>, 
                   PCA <span class="op">=</span> <span class="st">"raw"</span>, PCs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>, condition <span class="op">=</span> <span class="st">"groups"</span><span class="op">)</span>
<span class="fu">RFLOMICS</span><span class="fu">::</span><span class="fu">plotPCA</span><span class="op">(</span> <span class="va">RNAseq.SE</span>, 
                   PCA <span class="op">=</span> <span class="st">"norm"</span>, PCs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>, condition <span class="op">=</span> <span class="st">"groups"</span><span class="op">)</span>


<span class="fu">RFLOMICS</span><span class="fu">::</span><span class="fu">plotPCA</span><span class="op">(</span> <span class="va">RNAseq.raw.SE</span>, 
                   PCA <span class="op">=</span> <span class="st">"raw"</span>, PCs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>, condition <span class="op">=</span> <span class="st">"groups"</span><span class="op">)</span>
<span class="fu">RFLOMICS</span><span class="fu">::</span><span class="fu">plotPCA</span><span class="op">(</span> <span class="va">RNAseq.SE</span>, 
                   PCA <span class="op">=</span> <span class="st">"norm"</span>, PCs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>, condition <span class="op">=</span> <span class="st">"groups"</span><span class="op">)</span></code></pre></div>
<p><img src="RFLOMICS-command-lines_files/figure-html/plotPCA%20groups-1.png" width="700"><img src="RFLOMICS-command-lines_files/figure-html/plotPCA%20groups-2.png" width="700"><img src="RFLOMICS-command-lines_files/figure-html/plotPCA%20groups-3.png" width="700"><img src="RFLOMICS-command-lines_files/figure-html/plotPCA%20groups-4.png" width="700"><img src="RFLOMICS-command-lines_files/figure-html/plotPCA%20groups-5.png" width="700"><img src="RFLOMICS-command-lines_files/figure-html/plotPCA%20groups-6.png" width="700"></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">RFLOMICS</span><span class="fu">::</span><span class="fu">plotPCA</span><span class="op">(</span> <span class="va">RNAseq.raw.SE</span>, 
                   PCA <span class="op">=</span> <span class="st">"raw"</span>, PCs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, condition <span class="op">=</span> <span class="st">"imbibition"</span><span class="op">)</span>
<span class="fu">RFLOMICS</span><span class="fu">::</span><span class="fu">plotPCA</span><span class="op">(</span> <span class="va">RNAseq.SE</span>, 
                   PCA <span class="op">=</span> <span class="st">"norm"</span>, PCs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, condition <span class="op">=</span> <span class="st">"imbibition"</span><span class="op">)</span>


<span class="fu">RFLOMICS</span><span class="fu">::</span><span class="fu">plotPCA</span><span class="op">(</span> <span class="va">RNAseq.raw.SE</span>, 
                   PCA <span class="op">=</span> <span class="st">"raw"</span>, PCs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>, condition <span class="op">=</span> <span class="st">"imbibition"</span><span class="op">)</span>
<span class="fu">RFLOMICS</span><span class="fu">::</span><span class="fu">plotPCA</span><span class="op">(</span> <span class="va">RNAseq.SE</span>, 
                   PCA <span class="op">=</span> <span class="st">"norm"</span>, PCs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>, condition <span class="op">=</span> <span class="st">"imbibition"</span><span class="op">)</span>


<span class="fu">RFLOMICS</span><span class="fu">::</span><span class="fu">plotPCA</span><span class="op">(</span> <span class="va">RNAseq.raw.SE</span>, 
                   PCA <span class="op">=</span> <span class="st">"raw"</span>, PCs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>, condition <span class="op">=</span> <span class="st">"imbibition"</span><span class="op">)</span>
<span class="fu">RFLOMICS</span><span class="fu">::</span><span class="fu">plotPCA</span><span class="op">(</span> <span class="va">RNAseq.SE</span>, 
                   PCA <span class="op">=</span> <span class="st">"norm"</span>, PCs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>, condition <span class="op">=</span> <span class="st">"imbibition"</span><span class="op">)</span></code></pre></div>
<p><img src="RFLOMICS-command-lines_files/figure-html/plotPCA%20factor1-1.png" width="700"><img src="RFLOMICS-command-lines_files/figure-html/plotPCA%20factor1-2.png" width="700"><img src="RFLOMICS-command-lines_files/figure-html/plotPCA%20factor1-3.png" width="700"><img src="RFLOMICS-command-lines_files/figure-html/plotPCA%20factor1-4.png" width="700"><img src="RFLOMICS-command-lines_files/figure-html/plotPCA%20factor1-5.png" width="700"><img src="RFLOMICS-command-lines_files/figure-html/plotPCA%20factor1-6.png" width="700"></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#mvQCdata(object = FlomicsMultiAssay,data= paste0(data,".filtred") , PCA="norm", axis=5,pngFile= NULL) </span></code></pre></div>
<p>We can noticed that imbibition factor explained a very huge proportion of dataset variability: * PCA1 &amp; PCA2: DS vs LI  + 33.9 = 75 % * PCA3: EI vs LI+DS  % Whereas the temperature factor explain very few: * PCA4 and PCA5: low vs Elevated ,+ 3.6 = 6.6 %.</p>
</div>
<div id="differential-expression-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#differential-expression-analysis" class="anchor" aria-hidden="true"></a>Differential expression analysis</h2>
<ul>
<li>Run the differential analysis thanks to the ‘RunDiffAnalyis’ method which calls the glmFit function from the edgeR package and applied the Benjamini and Hochberg (BH) FDR correction. The designed statistical model is applied for each contrast. To filter diferentially expressed genes, the ‘FilterDiffAnalysis’ method <strong>has to be called</strong>. All the raw and filtered results are stored in the metadata list of the SummarizedExperiment Object.</li>
</ul>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># run diffAnalysis</span>
<span class="va">RNAseq.SE</span> <span class="op">&lt;-</span>
  <span class="fu">RunDiffAnalysis</span><span class="op">(</span> <span class="va">RNAseq.SE</span>, design <span class="op">=</span> <span class="va">Design</span>, contrastList <span class="op">=</span> <span class="va">Design</span><span class="op">@</span><span class="va">Contrasts.Sel</span><span class="op">$</span><span class="va">contrastName</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>, Adj.pvalue.method <span class="op">=</span> <span class="st">"BH"</span>, DiffAnalysisMethod <span class="op">=</span> <span class="st">"edgeRglmfit"</span>, clustermq <span class="op">=</span> <span class="cn">FALSE</span> <span class="op">)</span>
<span class="co">#&gt; [1] "[cmd] dge &lt;- edgeR::estimateGLMCommonDisp(dge, design=model_matrix)"</span>
<span class="co">#&gt; [1] "[cmd] dge &lt;- edgeR::estimateGLMTrendedDisp(dge, design=model_matrix)"</span>
<span class="co">#&gt; [1] "[cmd] dge &lt;- edgeR::estimateGLMTagwiseDisp(dge, design=model_matrix)"</span>
<span class="co">#&gt; [1] "[cmd] fit.f &lt;- edgeR::glmFit(dge,design=model_matrix)"</span>
<span class="co">#&gt; [1] "[cmd] apply model to each contrast"</span>
<span class="co"># results filering</span>
<span class="va">RNAseq.SE</span> <span class="op">&lt;-</span>
  <span class="fu">FilterDiffAnalysis</span><span class="op">(</span><span class="va">RNAseq.SE</span>, Adj.pvalue.cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span>
<span class="co">#&gt; [1] "(temperatureLow - temperatureElevated) in imbibitionLI"</span>
<span class="co">#&gt; [1] "H1"</span>
<span class="co">#&gt; [1] "(temperatureMedium - temperatureElevated) in imbibitionLI"</span>
<span class="co">#&gt; [1] "H2"</span>
<span class="co">#&gt; [1] "(temperatureMedium - temperatureLow) in imbibitionLI"</span>
<span class="co">#&gt; [1] "H3"</span>
<span class="co"># results access</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">RNAseq.SE</span><span class="op">@</span><span class="va">metadata</span><span class="op">)</span>
<span class="co">#&gt; [1] "omicType"         "Groups"           "rowSums.zero"     "FilteredFeatures"</span>
<span class="co">#&gt; [5] "FilteredOptions"  "FilteringOptions" "Normalization"    "PCAlist"         </span>
<span class="co">#&gt; [9] "DiffExpAnal"</span></code></pre></div>
<ul>
<li>Access to the number of up and down regulated genes.</li>
</ul>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tmpRes</span> <span class="op">&lt;-</span>
  <span class="va">RNAseq.SE</span><span class="op">@</span><span class="va">metadata</span><span class="op">[[</span><span class="st">"DiffExpAnal"</span><span class="op">]</span><span class="op">]</span>
<span class="va">DE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">DEup</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">DEdown</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span><span class="op">)</span>

<span class="kw">for</span> <span class="op">(</span><span class="va">i</span>  <span class="kw">in</span>  <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">tmpRes</span><span class="op">[[</span><span class="st">"contrasts"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">contrastName</span> <span class="op">&lt;-</span> <span class="va">tmpRes</span><span class="op">[[</span><span class="st">"contrasts"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">i</span>, <span class="op">]</span><span class="op">$</span><span class="va">contrastName</span>
  <span class="va">stats</span> <span class="op">&lt;-</span> <span class="va">tmpRes</span><span class="op">[[</span><span class="st">"stats"</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="va">contrastName</span><span class="op">]</span><span class="op">]</span>
  <span class="va">DE</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">stats</span><span class="op">$</span><span class="va">gDE</span>
  <span class="va">DEup</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"("</span>, <span class="va">stats</span><span class="op">$</span><span class="va">pgDEup</span>, <span class="st">" %)"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span>
  <span class="va">DEdown</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"("</span>, <span class="va">stats</span><span class="op">$</span><span class="va">pgDEdown</span>, <span class="st">" %)"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span>
<span class="op">}</span>


<span class="co"># knitr::kable(dplyr::bind_cols(</span>
<span class="co">#   tmpRes[["contrasts"]][, c("tag", "contrastName")],</span>
<span class="co">#   data.frame(</span>
<span class="co">#     "Nb DE" = DE,</span>
<span class="co">#     "Nb DEup" = DEup,</span>
<span class="co">#     "Nb DEdown" = DEdown</span>
<span class="co">#   )</span>
<span class="co"># ),</span>
<span class="co"># row.names = FALSE,</span>
<span class="co"># caption = NULL)</span>

<span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/bind.html" class="external-link">bind_cols</a></span><span class="op">(</span> <span class="va">tmpRes</span><span class="op">[[</span><span class="st">"contrasts"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tag"</span>, <span class="st">"contrastName"</span><span class="op">)</span><span class="op">]</span>,
                  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="st">"Nb DE"</span> <span class="op">=</span> <span class="va">DE</span>, <span class="st">"Nb DEup"</span> <span class="op">=</span> <span class="va">DEup</span>, <span class="st">"Nb DEdown"</span> <span class="op">=</span> <span class="va">DEdown</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;    tag                                              contrastName Nb.DE Nb.DEup</span>
<span class="co">#&gt; 1:  H1    (temperatureLow - temperatureElevated) in imbibitionLI  6651  (52 %)</span>
<span class="co">#&gt; 2:  H2 (temperatureMedium - temperatureElevated) in imbibitionLI  3097  (46 %)</span>
<span class="co">#&gt; 3:  H3      (temperatureMedium - temperatureLow) in imbibitionLI  3294  (37 %)</span>
<span class="co">#&gt;    Nb.DEdown</span>
<span class="co">#&gt; 1:    (48 %)</span>
<span class="co">#&gt; 2:    (54 %)</span>
<span class="co">#&gt; 3:    (63 %)</span></code></pre></div>
<p>We can see that there are more genes differentially expressed between imbibition modalities than between temperature modalities except for low vs moderated.</p>
<ul>
<li>Run the DiffAnal.plot function to get the MAplot and the Pvalue distribution. By example,</li>
</ul>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pDA</span> <span class="op">&lt;-</span>
  <span class="fu">DiffAnal.plot</span><span class="op">(</span><span class="va">RNAseq.SE</span>,
                <span class="va">tmpRes</span><span class="op">[[</span><span class="st">"contrasts"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span><span class="op">$</span><span class="va">contrastName</span> ,
                Adj.pvalue.cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></code></pre></div>
<ul>
<li>Check the MAplot:</li>
</ul>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pDA</span><span class="op">$</span><span class="va">MA.plot</span></code></pre></div>
<p><img src="RFLOMICS-command-lines_files/figure-html/MAplot-1.png" width="700"></p>
<ul>
<li>You <strong>must</strong> check the Pvalue distribution for each contrast.
<ul>
<li>If the distribution is uniform: it is ok. a few genes are DE and the FDR will find them.</li>
<li>If the distribution as a peak near 0 and then is uniform: it is also ok. The higher is the peak, the higher is the number of DE genes. FDR correction can be applied.</li>
<li>If the distribution has 2 peaks one near 0 and one near 1 or just one peak near 1, it is not a good news ! You have to understand what’s happened.</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pDA</span><span class="op">$</span><span class="va">Pvalue.hist</span></code></pre></div>
<p><img src="RFLOMICS-command-lines_files/figure-html/PvaluePlot-1.png" width="700"></p>
<ul>
<li>Draw a Venn diagramm of DEG for samples thanks to the UpSetR function.</li>
</ul>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">DEF_mat</span> <span class="op">&lt;-</span>
  <span class="va">RNAseq.SE</span><span class="op">@</span><span class="va">metadata</span><span class="op">$</span><span class="va">DiffExpAnal</span><span class="op">[[</span><span class="st">"mergeDEF"</span><span class="op">]</span><span class="op">]</span>
<span class="co">#UpSetR::upset(DEF_mat, sets = c("H22","H23","H24")) </span></code></pre></div>
<p>In this example, we focus on 3 contrasts which are the difference in imbibition modalities factor with an averaged response to temperature. Set Size is the number of DEGs per hypothesis. We can see that very few DEGs are specific to one contrast. 2573 DEGs are common to the DS-LI and DS-EI contrasts. 1631 DEGs are common to the ES-LI and DS-LI contrasts. 947 DEGs are common to EI_LI and DS-EI contrasts. 7007 DEG in the three contrasts.</p>
<div id="coexpression-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#coexpression-analysis" class="anchor" aria-hidden="true"></a>Coexpression analysis</h3>
<ul>
<li><p>A list of DEG associated to contrasts has to be constructed. Choice is given to merge (<strong>union</strong>) or intersect (<strong>intersection</strong>) the list of genes.</p></li>
<li><p>Co-expression analysis is done thanks to the coseq package and function with a set of optimal parameters which are the data transformation method: fixed to <strong>arcsin</strong> and the model: fixed to <strong>gaussian mixture models</strong>. The number of technical replicates to perform (<strong>iter</strong>) for each <strong>K</strong> which is the number of cluster into which the DEG genes have to be cut, can be first set to 0 to precise the range of K and then put to a minimum of 10 replicates. see the coseq vignette.</p></li>
</ul>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">DEG_list</span> <span class="op">&lt;-</span> <span class="fu">getDEGlist_for_coseqAnalysis</span><span class="op">(</span>
        matrix    <span class="op">=</span> <span class="va">RNAseq.SE</span><span class="op">@</span><span class="va">metadata</span><span class="op">$</span><span class="va">DiffExpAnal</span><span class="op">[[</span><span class="st">"mergeDEF"</span><span class="op">]</span><span class="op">]</span>,
        colnames  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"H1"</span>,<span class="st">"H2"</span>,<span class="st">"H3"</span><span class="op">)</span>,
        mergeType <span class="op">=</span> <span class="st">"union"</span><span class="op">)</span>

<span class="va">RNAseq.SE</span> <span class="op">&lt;-</span>  
<span class="fu">runCoExpression</span><span class="op">(</span><span class="va">RNAseq.SE</span>,
                                  geneList<span class="op">=</span><span class="va">DEG_list</span>,
                                  K<span class="op">=</span><span class="fl">2</span><span class="op">:</span><span class="fl">4</span>, 
                                  replicate <span class="op">=</span> <span class="fl">5</span>,
                                  model  <span class="op">=</span> <span class="st">"normal"</span>,
                                  transformation<span class="op">=</span> <span class="st">"arcsin"</span>, 
                                  normFactors<span class="op">=</span><span class="st">"TMM"</span>,
                                  nameList<span class="op">=</span><span class="st">"List1"</span>, 
                                  merge<span class="op">=</span> <span class="st">"union"</span>,
                                  GaussianModel <span class="op">=</span> <span class="st">"Gaussian_pk_Lk_Ck"</span><span class="op">)</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; coseq analysis: Normal approach &amp; arcsin transformation</span>
<span class="co">#&gt; K = 2 to 2 </span>
<span class="co">#&gt; Use seed argument in coseq for reproducible results.</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; Running K = 2 ...</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; coseq analysis: Normal approach &amp; arcsin transformation</span>
<span class="co">#&gt; K = 3 to 3 </span>
<span class="co">#&gt; Use seed argument in coseq for reproducible results.</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; Running K = 3 ...</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; coseq analysis: Normal approach &amp; arcsin transformation</span>
<span class="co">#&gt; K = 4 to 4 </span>
<span class="co">#&gt; Use seed argument in coseq for reproducible results.</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; Running K = 4 ...</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; coseq analysis: Normal approach &amp; arcsin transformation</span>
<span class="co">#&gt; K = 2 to 2 </span>
<span class="co">#&gt; Use seed argument in coseq for reproducible results.</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; Running K = 2 ...</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; coseq analysis: Normal approach &amp; arcsin transformation</span>
<span class="co">#&gt; K = 3 to 3 </span>
<span class="co">#&gt; Use seed argument in coseq for reproducible results.</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; Running K = 3 ...</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; coseq analysis: Normal approach &amp; arcsin transformation</span>
<span class="co">#&gt; K = 4 to 4 </span>
<span class="co">#&gt; Use seed argument in coseq for reproducible results.</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; Running K = 4 ...</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; coseq analysis: Normal approach &amp; arcsin transformation</span>
<span class="co">#&gt; K = 2 to 2 </span>
<span class="co">#&gt; Use seed argument in coseq for reproducible results.</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; Running K = 2 ...</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; coseq analysis: Normal approach &amp; arcsin transformation</span>
<span class="co">#&gt; K = 3 to 3 </span>
<span class="co">#&gt; Use seed argument in coseq for reproducible results.</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; Running K = 3 ...</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; coseq analysis: Normal approach &amp; arcsin transformation</span>
<span class="co">#&gt; K = 4 to 4 </span>
<span class="co">#&gt; Use seed argument in coseq for reproducible results.</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; Running K = 4 ...</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; coseq analysis: Normal approach &amp; arcsin transformation</span>
<span class="co">#&gt; K = 2 to 2 </span>
<span class="co">#&gt; Use seed argument in coseq for reproducible results.</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; Running K = 2 ...</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; coseq analysis: Normal approach &amp; arcsin transformation</span>
<span class="co">#&gt; K = 3 to 3 </span>
<span class="co">#&gt; Use seed argument in coseq for reproducible results.</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; Running K = 3 ...</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; coseq analysis: Normal approach &amp; arcsin transformation</span>
<span class="co">#&gt; K = 4 to 4 </span>
<span class="co">#&gt; Use seed argument in coseq for reproducible results.</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; Running K = 4 ...</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; coseq analysis: Normal approach &amp; arcsin transformation</span>
<span class="co">#&gt; K = 2 to 2 </span>
<span class="co">#&gt; Use seed argument in coseq for reproducible results.</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; Running K = 2 ...</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; coseq analysis: Normal approach &amp; arcsin transformation</span>
<span class="co">#&gt; K = 3 to 3 </span>
<span class="co">#&gt; Use seed argument in coseq for reproducible results.</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; Running K = 3 ...</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; coseq analysis: Normal approach &amp; arcsin transformation</span>
<span class="co">#&gt; K = 4 to 4 </span>
<span class="co">#&gt; Use seed argument in coseq for reproducible results.</span>
<span class="co">#&gt; ****************************************</span>
<span class="co">#&gt; Running K = 4 ...</span>
<span class="co">#&gt; [1] 3</span></code></pre></div>
<p>In this example, we took the union of three DEG lists.</p>
<ul>
<li>Plot likelihood</li>
</ul>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">RNAseq.SE</span><span class="op">@</span><span class="va">metadata</span><span class="op">$</span><span class="va">CoExpAnal</span><span class="op">$</span><span class="va">plots</span><span class="op">$</span><span class="va">ICL</span></code></pre></div>
<p><img src="RFLOMICS-command-lines_files/figure-html/plotCoseq1-1.png" width="700"></p>
<p>The best number of clusters (K) into which DEG list’s union have to be cut is 10.</p>
<ul>
<li>Plot Expression profiles</li>
</ul>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">RNAseq.SE</span><span class="op">@</span><span class="va">metadata</span><span class="op">$</span><span class="va">CoExpAnal</span><span class="op">$</span><span class="va">plots</span><span class="op">$</span><span class="va">boxplots</span></code></pre></div>
<p><img src="RFLOMICS-command-lines_files/figure-html/plotCoexp-1.png" width="700"></p>
</div>
</div>
<div id="annotation-and-enrichment-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#annotation-and-enrichment-analysis" class="anchor" aria-hidden="true"></a>Annotation and Enrichment analysis</h2>
<p>To perform annotation and enrichment analysis, we need Reference file (see <a href="/private/var/folders/mn/95p840s571b4mgdjr6n330fr0000gp/T/RtmpEbqx08/temp_libpath14173db9c5ef/RFLOMICS//docs/articles/RFLOMICS-input-data.html">rflomics input data</a>.</p>
<p>Then we select a lists of DGEs and clusters to annotate. By default the annotation whill be done on all setected DGE lists and clusters.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ListNames.diff</span>  <span class="op">&lt;-</span> <span class="va">RNAseq.SE</span><span class="op">@</span><span class="va">metadata</span><span class="op">$</span><span class="va">DiffExpAnal</span><span class="op">[[</span><span class="st">"contrasts"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">contrastName</span>
<span class="va">ListNames.coseq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">RNAseq.SE</span><span class="op">@</span><span class="va">metadata</span><span class="op">$</span><span class="va">CoExpAnal</span><span class="op">[[</span><span class="st">"clusters"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<p>Run gene enrichment with hypergenometric method.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">RNAseq.SE</span> <span class="op">&lt;-</span> <span class="fu">runAnnotationEnrichment</span><span class="op">(</span> <span class="va">RNAseq.SE</span>, annotation <span class="op">=</span> <span class="va">annot.genes</span>, 
                                      DiffListNames <span class="op">=</span> <span class="va">ListNames.diff</span>, 
                                      CoExpListNames<span class="op">=</span> <span class="va">ListNames.coseq</span>,
                                      alpha <span class="op">=</span> <span class="fl">0.01</span>, 
                                      probaMethod <span class="op">=</span> <span class="st">"hypergeometric"</span><span class="op">)</span></code></pre></div>
<p>Plot the over-represented GO terms for each list.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R">  <span class="co"># top based on pvalue</span>
  <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Enrichment.plot.html">Enrichment.plot</a></span><span class="op">(</span><span class="va">RNAseq.SE</span>, Over_Under<span class="op">=</span><span class="st">"overrepresented"</span>, top <span class="op">=</span> <span class="fl">50</span>, listNames<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cluster.4"</span>, <span class="st">"cluster.3"</span><span class="op">)</span><span class="op">)</span>

  <span class="va">p</span>
<span class="co">#&gt; $cluster.4</span></code></pre></div>
<p><img src="RFLOMICS-command-lines_files/figure-html/pvalue.enrichment.plot%20overrepresented-1.png" width="700"></p>
<pre><code>#&gt; 
#&gt; $cluster.3</code></pre>
<p><img src="RFLOMICS-command-lines_files/figure-html/pvalue.enrichment.plot%20overrepresented-2.png" width="700"></p>
<p>Plot the over-represented GO terms for each list.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R">  <span class="co"># top based on pvalue</span>
  <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Enrichment.plot.html">Enrichment.plot</a></span><span class="op">(</span><span class="va">RNAseq.SE</span>, Over_Under<span class="op">=</span><span class="st">"underrepresented"</span>, top <span class="op">=</span> <span class="fl">50</span>, listNames<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cluster.4"</span>, <span class="st">"cluster.3"</span><span class="op">)</span><span class="op">)</span>

  <span class="va">p</span>
<span class="co">#&gt; $cluster.4</span></code></pre></div>
<p><img src="RFLOMICS-command-lines_files/figure-html/pvalue.enrichment.plot%20underrepresented-1.png" width="700"></p>
<pre><code>#&gt; 
#&gt; $cluster.3</code></pre>
<p><img src="RFLOMICS-command-lines_files/figure-html/pvalue.enrichment.plot%20underrepresented-2.png" width="700"></p>
</div>
<div id="section" class="section level2">
<h2></h2>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">FlomicsMultiAssay</span><span class="op">@</span><span class="va">ExperimentList</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">data</span>, <span class="st">".filtred"</span><span class="op">)</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">RNAseq.SE</span></code></pre></div>
</div>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor" aria-hidden="true"></a>Session information:</h1>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/sessioninfo/man/session_info.html" class="external-link">session_info</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; ─ Session info ───────────────────────────────────────────────────────────────</span>
<span class="co">#&gt;  setting  value                       </span>
<span class="co">#&gt;  version  R version 4.1.0 (2021-05-18)</span>
<span class="co">#&gt;  os       macOS Mojave 10.14.6        </span>
<span class="co">#&gt;  system   x86_64, darwin17.0          </span>
<span class="co">#&gt;  ui       X11                         </span>
<span class="co">#&gt;  language (EN)                        </span>
<span class="co">#&gt;  collate  fr_FR.UTF-8                 </span>
<span class="co">#&gt;  ctype    fr_FR.UTF-8                 </span>
<span class="co">#&gt;  tz       Europe/Paris                </span>
<span class="co">#&gt;  date     2021-10-06                  </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ─ Packages ───────────────────────────────────────────────────────────────────</span>
<span class="co">#&gt;  package              * version    date       lib</span>
<span class="co">#&gt;  annotate               1.70.0     2021-05-19 [2]</span>
<span class="co">#&gt;  AnnotationDbi          1.54.1     2021-06-08 [2]</span>
<span class="co">#&gt;  assertthat             0.2.1      2019-03-21 [2]</span>
<span class="co">#&gt;  bayesm                 3.1-4      2019-10-15 [2]</span>
<span class="co">#&gt;  Biobase                2.52.0     2021-05-19 [2]</span>
<span class="co">#&gt;  BiocGenerics           0.38.0     2021-05-19 [2]</span>
<span class="co">#&gt;  BiocManager            1.30.16    2021-06-15 [2]</span>
<span class="co">#&gt;  BiocParallel           1.26.2     2021-08-22 [2]</span>
<span class="co">#&gt;  BiocStyle            * 2.20.2     2021-06-17 [2]</span>
<span class="co">#&gt;  Biostrings             2.60.2     2021-08-05 [2]</span>
<span class="co">#&gt;  bit                    4.0.4      2020-08-04 [2]</span>
<span class="co">#&gt;  bit64                  4.0.5      2020-08-30 [2]</span>
<span class="co">#&gt;  bitops                 1.0-7      2021-04-24 [2]</span>
<span class="co">#&gt;  blob                   1.2.2      2021-07-23 [2]</span>
<span class="co">#&gt;  bookdown               0.24       2021-09-02 [2]</span>
<span class="co">#&gt;  bslib                  0.3.0      2021-09-02 [2]</span>
<span class="co">#&gt;  cachem                 1.0.6      2021-08-19 [2]</span>
<span class="co">#&gt;  callr                  3.7.0      2021-04-20 [2]</span>
<span class="co">#&gt;  capushe                1.1.1      2016-04-19 [2]</span>
<span class="co">#&gt;  class                  7.3-19     2021-05-03 [2]</span>
<span class="co">#&gt;  cli                    3.0.1      2021-07-17 [2]</span>
<span class="co">#&gt;  cluster                2.1.2      2021-04-17 [2]</span>
<span class="co">#&gt;  colorspace             2.0-2      2021-06-24 [2]</span>
<span class="co">#&gt;  compositions           2.0-2      2021-07-14 [2]</span>
<span class="co">#&gt;  corrplot               0.90       2021-06-30 [2]</span>
<span class="co">#&gt;  coseq                  1.16.0     2021-05-19 [2]</span>
<span class="co">#&gt;  crayon                 1.4.1      2021-02-08 [2]</span>
<span class="co">#&gt;  data.table             1.14.0     2021-02-21 [2]</span>
<span class="co">#&gt;  DBI                    1.1.1      2021-01-15 [2]</span>
<span class="co">#&gt;  DelayedArray           0.18.0     2021-05-19 [2]</span>
<span class="co">#&gt;  DEoptimR               1.0-9      2021-05-24 [2]</span>
<span class="co">#&gt;  desc                   1.4.0      2021-09-28 [2]</span>
<span class="co">#&gt;  DESeq2                 1.32.0     2021-05-19 [2]</span>
<span class="co">#&gt;  devtools               2.4.2      2021-06-07 [2]</span>
<span class="co">#&gt;  digest                 0.6.28     2021-09-23 [2]</span>
<span class="co">#&gt;  dplyr                  1.0.7      2021-06-18 [2]</span>
<span class="co">#&gt;  DT                     0.19       2021-09-02 [2]</span>
<span class="co">#&gt;  e1071                  1.7-9      2021-09-16 [2]</span>
<span class="co">#&gt;  edgeR                  3.34.1     2021-09-09 [2]</span>
<span class="co">#&gt;  ellipsis               0.3.2      2021-04-29 [2]</span>
<span class="co">#&gt;  evaluate               0.14       2019-05-28 [2]</span>
<span class="co">#&gt;  FactoMineR             2.4        2020-12-11 [2]</span>
<span class="co">#&gt;  fansi                  0.5.0      2021-05-25 [2]</span>
<span class="co">#&gt;  farver                 2.1.0      2021-02-28 [2]</span>
<span class="co">#&gt;  fastmap                1.1.0      2021-01-25 [2]</span>
<span class="co">#&gt;  flashClust             1.01-2     2012-08-21 [2]</span>
<span class="co">#&gt;  fs                     1.5.0      2020-07-31 [2]</span>
<span class="co">#&gt;  genefilter             1.74.0     2021-05-19 [2]</span>
<span class="co">#&gt;  geneplotter            1.70.0     2021-05-19 [2]</span>
<span class="co">#&gt;  generics               0.1.0      2020-10-31 [2]</span>
<span class="co">#&gt;  GenomeInfoDb           1.28.4     2021-09-09 [2]</span>
<span class="co">#&gt;  GenomeInfoDbData       1.2.6      2021-07-12 [2]</span>
<span class="co">#&gt;  GenomicRanges          1.44.0     2021-05-19 [2]</span>
<span class="co">#&gt;  ggplot2              * 3.3.5      2021-06-25 [2]</span>
<span class="co">#&gt;  ggrepel                0.9.1      2021-01-15 [2]</span>
<span class="co">#&gt;  glue                   1.4.2      2020-08-27 [2]</span>
<span class="co">#&gt;  gtable                 0.3.0      2019-03-25 [2]</span>
<span class="co">#&gt;  highr                  0.9        2021-04-16 [2]</span>
<span class="co">#&gt;  htmltools              0.5.2      2021-08-25 [2]</span>
<span class="co">#&gt;  htmlwidgets            1.5.4      2021-09-08 [2]</span>
<span class="co">#&gt;  HTSCluster             2.0.8      2016-05-26 [2]</span>
<span class="co">#&gt;  HTSFilter              1.32.0     2021-05-19 [2]</span>
<span class="co">#&gt;  httr                   1.4.2      2020-07-20 [2]</span>
<span class="co">#&gt;  IRanges                2.26.0     2021-05-19 [2]</span>
<span class="co">#&gt;  jquerylib              0.1.4      2021-04-26 [2]</span>
<span class="co">#&gt;  jsonlite               1.7.2      2020-12-09 [2]</span>
<span class="co">#&gt;  KEGGREST               1.32.0     2021-05-19 [2]</span>
<span class="co">#&gt;  knitr                  1.34       2021-09-09 [2]</span>
<span class="co">#&gt;  labeling               0.4.2      2020-10-20 [2]</span>
<span class="co">#&gt;  lattice                0.20-44    2021-05-02 [2]</span>
<span class="co">#&gt;  leaps                  3.1        2020-01-16 [2]</span>
<span class="co">#&gt;  lifecycle              1.0.1      2021-09-24 [2]</span>
<span class="co">#&gt;  limma                  3.48.3     2021-08-10 [2]</span>
<span class="co">#&gt;  locfit                 1.5-9.4    2020-03-25 [2]</span>
<span class="co">#&gt;  magrittr               2.0.1      2020-11-17 [2]</span>
<span class="co">#&gt;  MASS                   7.3-54     2021-05-03 [2]</span>
<span class="co">#&gt;  Matrix                 1.3-4      2021-06-01 [2]</span>
<span class="co">#&gt;  MatrixGenerics         1.4.3      2021-08-26 [2]</span>
<span class="co">#&gt;  matrixStats            0.61.0     2021-09-17 [2]</span>
<span class="co">#&gt;  memoise                2.0.0      2021-01-26 [2]</span>
<span class="co">#&gt;  MultiAssayExperiment   1.18.0     2021-05-19 [2]</span>
<span class="co">#&gt;  munsell                0.5.0      2018-06-12 [2]</span>
<span class="co">#&gt;  mvtnorm                1.1-2      2021-06-07 [2]</span>
<span class="co">#&gt;  pillar                 1.6.3      2021-09-26 [2]</span>
<span class="co">#&gt;  pkgbuild               1.2.0      2020-12-15 [2]</span>
<span class="co">#&gt;  pkgconfig              2.0.3      2019-09-22 [2]</span>
<span class="co">#&gt;  pkgdown                1.6.1.9001 2021-09-28 [2]</span>
<span class="co">#&gt;  pkgload                1.2.2      2021-09-11 [2]</span>
<span class="co">#&gt;  plotrix                3.8-2      2021-09-08 [2]</span>
<span class="co">#&gt;  plyr                   1.8.6      2020-03-03 [2]</span>
<span class="co">#&gt;  png                    0.1-7      2013-12-03 [2]</span>
<span class="co">#&gt;  prettyunits            1.1.1      2020-01-24 [2]</span>
<span class="co">#&gt;  processx               3.5.2      2021-04-30 [2]</span>
<span class="co">#&gt;  proxy                  0.4-26     2021-06-07 [2]</span>
<span class="co">#&gt;  ps                     1.6.0      2021-02-28 [2]</span>
<span class="co">#&gt;  purrr                  0.3.4      2020-04-17 [2]</span>
<span class="co">#&gt;  R6                     2.5.1      2021-08-19 [2]</span>
<span class="co">#&gt;  ragg                   1.1.3      2021-06-09 [2]</span>
<span class="co">#&gt;  RColorBrewer           1.1-2      2014-12-07 [2]</span>
<span class="co">#&gt;  Rcpp                   1.0.7      2021-07-07 [2]</span>
<span class="co">#&gt;  RCurl                  1.98-1.5   2021-09-17 [2]</span>
<span class="co">#&gt;  remotes                2.4.0      2021-06-02 [2]</span>
<span class="co">#&gt;  reshape2               1.4.4      2020-04-09 [2]</span>
<span class="co">#&gt;  RFLOMICS             * 0.0.0.9000 2021-10-06 [1]</span>
<span class="co">#&gt;  rlang                  0.4.11     2021-04-30 [2]</span>
<span class="co">#&gt;  rmarkdown              2.11       2021-09-14 [2]</span>
<span class="co">#&gt;  Rmixmod                2.1.5      2020-06-14 [2]</span>
<span class="co">#&gt;  robustbase             0.93-8     2021-06-02 [2]</span>
<span class="co">#&gt;  rprojroot              2.0.2      2020-11-15 [2]</span>
<span class="co">#&gt;  RSQLite                2.2.8      2021-08-21 [2]</span>
<span class="co">#&gt;  rstudioapi             0.13       2020-11-12 [2]</span>
<span class="co">#&gt;  S4Vectors              0.30.0     2021-05-19 [2]</span>
<span class="co">#&gt;  sass                   0.4.0      2021-05-12 [2]</span>
<span class="co">#&gt;  scales                 1.1.1      2020-05-11 [2]</span>
<span class="co">#&gt;  scatterplot3d          0.3-41     2018-03-14 [2]</span>
<span class="co">#&gt;  sessioninfo            1.1.1      2018-11-05 [2]</span>
<span class="co">#&gt;  stringi                1.7.4      2021-08-25 [2]</span>
<span class="co">#&gt;  stringr                1.4.0      2019-02-10 [2]</span>
<span class="co">#&gt;  SummarizedExperiment   1.22.0     2021-05-19 [2]</span>
<span class="co">#&gt;  survival               3.2-13     2021-08-24 [2]</span>
<span class="co">#&gt;  systemfonts            1.0.2      2021-05-11 [2]</span>
<span class="co">#&gt;  tensorA                0.36.2     2020-11-19 [2]</span>
<span class="co">#&gt;  testthat               3.0.4      2021-07-01 [2]</span>
<span class="co">#&gt;  textshaping            0.3.5      2021-06-09 [2]</span>
<span class="co">#&gt;  tibble                 3.1.4      2021-08-25 [2]</span>
<span class="co">#&gt;  tidyr                  1.1.3      2021-03-03 [2]</span>
<span class="co">#&gt;  tidyselect             1.1.1      2021-04-30 [2]</span>
<span class="co">#&gt;  usethis                2.0.1      2021-02-10 [2]</span>
<span class="co">#&gt;  utf8                   1.2.2      2021-07-24 [2]</span>
<span class="co">#&gt;  vctrs                  0.3.8      2021-04-29 [2]</span>
<span class="co">#&gt;  withr                  2.4.2      2021-04-18 [2]</span>
<span class="co">#&gt;  xfun                   0.26       2021-09-14 [2]</span>
<span class="co">#&gt;  XML                    3.99-0.8   2021-09-17 [2]</span>
<span class="co">#&gt;  xtable                 1.8-4      2019-04-21 [2]</span>
<span class="co">#&gt;  XVector                0.32.0     2021-05-19 [2]</span>
<span class="co">#&gt;  yaml                   2.2.1      2020-02-01 [2]</span>
<span class="co">#&gt;  zlibbioc               1.38.0     2021-05-19 [2]</span>
<span class="co">#&gt;  source                        </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  Github (r-lib/pkgdown@6328892)</span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  local                         </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt;  CRAN (R 4.1.0)                </span>
<span class="co">#&gt;  Bioconductor                  </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [1] /private/var/folders/mn/95p840s571b4mgdjr6n330fr0000gp/T/RtmpEbqx08/temp_libpath14173db9c5ef</span>
<span class="co">#&gt; [2] /Library/Frameworks/R.framework/Versions/4.1/Resources/library</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by First Last.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link external-link">pkgdown</a> 1.6.1.9001.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
