<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>RFLOMICS • RFLOMICS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="RFLOMICS">
<meta property="og:description" content="RFLOMICS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RFLOMICS</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/RFLOMICS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/RFLOMICS-input-data.html">RFLOMICS input data</a>
    </li>
    <li>
      <a href="../articles/RFLOMICS-configure-cluster-access.html">RFLOMICS remote computing</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://forgemia.inra.fr/flomics/rflomics" class="external-link">
    <span class="fa fa-gitlab"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>RFLOMICS</h1>
            
      
      
      <div class="hidden name"><code>RFLOMICS.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p style="text-align: justify;">
      Here, we will walk through a multi-omics data analysis with the
      RFLOMICS package. We will see how to set up the statistical
      framework of the designed protocol: ie translate biological
      condition comparisons into contrasts, perform quality controls and
      statistical analysis of each omics dataset: ie. find
      differentially expressed genes/proteins/metabolites (DEF) per
      contrasts, identify clusters of co-expressed genes and perform a
      gene set enrichment analysis from lists of DEF or from clusters of
      co-expressed genes/proteins/metabolites.
      </p>
    </div>
    
<p><br><br></p>
<div class="section level2">
<h2 id="run-rflomics">1. Run RFLOMICS<a class="anchor" aria-label="anchor" href="#run-rflomics"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">library</span><span class="er">(</span><span class="ex">RFLOMICS</span><span class="kw">)</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">RFLOMICS::runExample()</span></span></code></pre></div>
<p>On the left side of interface, the different tab menu will appear as
the analysis progresses to guide the user.</p>
</div>
<div class="section level2">
<h2 id="load-data">2. Load Data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h2>
<p>We start by setting a project name (it is mandatory). Then we load
the experimental design and the different omics datasets.</p>
<div class="section level3">
<h3 id="dataset-example-description">2-1 Dataset example description:<a class="anchor" aria-label="anchor" href="#dataset-example-description"></a>
</h3>
<p style="text-align: justify;">
The data used for this vignette have been provided by Pr. Loic Rajjou
and Gwendal Cueff. They are included in the inst/ExampleFiles/ecoseed
directory of the package. Briefly, A. thaliana’s transcriptome,
metabolome and proteome have been obtained in the context of the study
of seed germination and vigor. In particular, the authors were
interested in deciphering key entities involved in response to
environmental stresses (on the mother plant): influences of temperature
(high, medium and low) and imbibition stage (Dry: DI, early imbibition:
EI and late imbibition: LI). See(<a href="https://www.uibk.ac.at/botany/ecoseed/home/" class="external-link uri">https://www.uibk.ac.at/botany/ecoseed/home/</a>)
</p>
</div>
<div class="section level3">
<h3 id="load-experimental-design">2-2 Load Experimental Design<a class="anchor" aria-label="anchor" href="#load-experimental-design"></a>
</h3>
<p>The experimental design file contains information about the readiness
status of each sample (see link : vignette input ). As soon as this file
is loaded, we can define the type (Biological factor : Bio; batch factor
: batch; metadata information : meta) and reference level of each design
factor.</p>
<ul>
<li><p><em>It is required to have minimum 1 biological factor (maximum 3
biological factors) and 1 batch factor. The order of the columns does
not matter.</em></p></li>
<li><p><em>It possible to exclude samples from design. These samples
will be excluded from the analysis even if they are present in the omics
dataset matrix.</em></p></li>
</ul>
</div>
<div class="section level3">
<h3 id="load-omics-data">2-3 Load Omics data<a class="anchor" aria-label="anchor" href="#load-omics-data"></a>
</h3>
<p style="text-align: justify;">
</p>
<p>For this tutorial we provide 3 datasets: RNAseq data (read counts),
proteomics data (abundance of proteins), and metabolomics data
(intensity of metabolites) (see link : vignette input). We must specify
for each dataset its type of omics. We can give a suitable name to our
dataset. By Default, they will be called respectively (set1, set2,
set3).</p>
As soon as the data are loaded the <strong>RFlomics
MultiAssayExperiment</strong> object is created (see link : data
management). The upset representation provides details about sample
intersection of different datasets.

<p><img src="Images/Capture_Load-Data.png" style="width:80.0%"></p>
</div>
<div class="section level3">
<h3 id="completeness-check">2-4 Completeness check<a class="anchor" aria-label="anchor" href="#completeness-check"></a>
</h3>
<p style="text-align: justify;">
</p>
<p>The design must be complete and as possible balanced. A
<strong>Complete design (mandatory)</strong> mean that all possible
combinations of biological factor’s level (called groups) are presents.
A <strong>Balanced design (recommended)</strong> has an equal number of
observations (replicates) for all groups.</p>
This constraint is checked and represented as graph (each square
represent 1 group and each color level indicate the number of
observations).

<p><img src="Images/completBalanced.png" width="50%"><img src="Images/completNonBalanced.png" width="50%"></p>
</div>
</div>
<div class="section level2">
<h2 id="set-the-statistical-framework">3. Set the statistical framework<a class="anchor" aria-label="anchor" href="#set-the-statistical-framework"></a>
</h2>
<p style="text-align: justify;">
</p>
<p>The interface provides models written from the simplest one to the
most complete one. Only the second orders interaction terms between the
biological factors will appear. Batch factor will never appear in
interaction terms. In our case we choose the complete model.</p>
<p>All ‘a priori’ hypothesis (contrast) calculated from the chosen model
are displayed (see link vignette contrast). The user must select those
that correspond to the biological issues.</p>
<p>In our case we are studying the effect of temperature on seed
germination (transition from the state of dormancy to the state of
germination vigor). For that we select 3 average contrasts (see
image):</p>
This statistical framework will be applied on all loaded datasets.

<p><img src="Images/Capture_Experimental-Design.png" style="width:80.0%"></p>
</div>
<div class="section level2">
<h2 id="omics-data-analysis">4. Omics data analysis<a class="anchor" aria-label="anchor" href="#omics-data-analysis"></a>
</h2>
<p>Once the contrasts are chosen, 3 Item menus appear in the side bar
menu that correspond to the 3 loaded omics types.</p>
<p>For each dataset, different steps of data analysis are proposed as a
tab panel. These analysis steps must be performed sequentially. It is
possible to switch between datasets. However, you cannot perform a new
task until the previous one is completed. A progress bar will be display
indicating the progress status.</p>
<div class="section level3">
<h3 id="data-exploratory">4-1 Data Exploratory<a class="anchor" aria-label="anchor" href="#data-exploratory"></a>
</h3>
<p>In this tab panel we access a data quality control screen and we
perform the appropriate data filtering and processing according to the
type of omics.</p>
<p>By default, features with 0 count in all conditions are removed from
the data.</p>
<p>Specifically for each data we have hte possibility to exclude samples
or to select a subset of samples to analyze, provided the completeness
condition is met.</p>
<p>The effect of data filtering or sample removing step can be explored
thanks to many QC graphs. For each QC, a graph is build on each raw data
and filtered data to allow the comparison.</p>
<div class="section level4">
<h4 id="rnaseq-data">RNAseq data<a class="anchor" aria-label="anchor" href="#rnaseq-data"></a>
</h4>
<p>For RNAseq data, low expressed genes are filtered such that only
genes whose cpm expression is greater than <strong>“CPM cutoff”</strong>
in x samples are kept. The value of x is given by <strong>“Filtering
strategy”</strong> (NbOfsample_over_cpm &lt;= NbConditions). By default
the data are filtered based on NbCondition strategy with CPM cutoff
equal to 5. In this example, there are 9 conditions. So, the number of
genes with a CPM less than 5 in at least 9 samples are removed.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># geneToKeep &lt;- counts[rowSums(edgeR::cpm(counts) &gt;= CPM_Cutoff) &gt;= x, ]</span></span></code></pre></div>
<p>To correct for differences in sequencing depths (library size),
RFlomics proposes one method for normalization, <strong>“TMM
method”</strong> from edgeR package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># dge &lt;- edgeR::DGEList(counts=counts, group=groups)</span></span>
<span><span class="co"># dge &lt;- edgeR::calcNormFactors(dge, method="TMM")</span></span></code></pre></div>
<p>For RNAses data the QC graphs are:</p>
<ul>
<li>The library size distribution which</li>
<li>Data Summary:</li>
<li>Distribution:</li>
<li>Distribution density</li>
<li>Principal component analysis</li>
</ul>
<p><img src="Images/capture_data_process_rnaseq.png" style="width:80.0%"></p>
</div>
<div class="section level4">
<h4 id="proteomics-or-metabolomics-data">Proteomics or Metabolomics data<a class="anchor" aria-label="anchor" href="#proteomics-or-metabolomics-data"></a>
</h4>
</div>
</div>
<div class="section level3">
<h3 id="differential-expression-analysis">4-2 Differential expression analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis"></a>
</h3>
<p>In this tab panel, we can run the differential expression analysis.
The analysis is run by contrast.<br>
For RNAseq data, the <strong>glmFit()</strong> function from the
<strong>edgeR</strong> package is used whereas for proteomic and
metabolomic data, the <strong>lmFit()</strong> function from
<strong>limma</strong> packages is used. The ability to run analysis on
a cluster is offered by sliding the cluster button (see
RFLOMICS-configure-cluster-access vignette)</p>
<p>Results will appear in two components: a scrolling menu will give the
results of the analysis by contrat and, at the bottom of the panel, the
intersecting sets of the lists of DEG is given thanks to the
<strong>UpSetR()</strong> function.</p>
<p>Each results has to bee validated:</p>
<ul>
<li>You <strong>have to</strong> to check the Pvalue distribution for
each contrast and validate it (<strong>ok checkbox</strong>).
<ul>
<li>If the distribution is uniform: it is ok. a few genes are DE and the
FDR will find them.</li>
<li>If the distribution as a peak near 0 and then is uniform: it is also
ok. The higher the peak, the higher the number of DE genes. FDR
correction can be applied.</li>
<li>If the distribution has 2 peaks one near 0 and one near 1 or just
one peak near 1, it is not a good news! You have to understand what’s
happened.</li>
</ul>
</li>
</ul>
<div class="figure">
<embed src="Images/Pvalue_distrib.pdf" style="width:80.0%"></embed><p class="caption">Pvalue-Distrib</p>
</div>
<ul>
<li><p>You can filter differentially expressed
genes/proteins/metabolites either by fold change (<strong>|FC|</strong>)
or by False Discovery Rate (BH) cut-off. Graphs will be automatically
updated.</p></li>
<li><p>You have to <strong>validate</strong> to save the cut-off
thresholds but also the contrasts’ result selection and pass to the
co-expression analysis or to the annotation enrichment step.</p></li>
</ul>
<p><img src="Images/capture_diff_analysis.png" style="width:80.0%"></p>
</div>
<div class="section level3">
<h3 id="coexpression-analysis">4-3 Coexpression analysis<a class="anchor" aria-label="anchor" href="#coexpression-analysis"></a>
</h3>
<ul>
<li><p>In this tab panel, choice is given to merge
(<strong>union</strong>) or intersect (<strong>intersection</strong>)
the lists of differentially expressed genes/proteins/metabolites
associated to contrasts.</p></li>
<li>
<p>Co-expression analysis is done using to the
<strong>coseq</strong> package with a set of optimal parameters.</p>
<ul>
<li>For <strong>RNAseq</strong>, data transformation method is set to
the <strong>arcsin</strong> function and normalization method to
<strong>TMM</strong>. No scaling is done for these data.
<strong>gaussian mixture model</strong> is used to decipher the
different profiles of gene expression with a set of parameters to
estimate fixed to (Gaussian_pK_Lk_Ck).</li>
<li>For <strong>proteomics/metabolomics</strong> data, scaling is done
on proteins/metabolites. Another sets of parameters estimation is also
proposed for the modeling of gene expression profile:
(Gaussian_pK_Lk_Bk). It has to be used, if the other one doesn’t
fit.</li>
</ul>
</li>
</ul>
<p>The number of technical replicates to perform (<strong>iter</strong>)
for each <strong>K</strong> which is the number of expressions’ profile
into which the DEG have to be cut, can be first set to 2 to precise the
range of K and then put to a minimum of 20 replicates per K.</p>
<p>The ability to run the clustering analysis on a cluster is offered by
sliding the Cluster button (see Rflomics_configure-cluster-access).</p>
<ul>
<li>
<p>Results description (see <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/coseq/inst/doc/coseq.html" class="external-link">CoSeq</a>)</p>
<ul>
<li><p>Results can be explored with several plots proposed by
<strong>coseq</strong>. The ICL graph has been slightly modified to show
all the range of the ICL values for a given K.</p></li>
<li><p>A table of jobs summary is also given. It groups error messages
by K.</p></li>
</ul>
</li>
</ul>
<p><img src="Images/capture_coexpression.png" style="width:80.0%"></p>
</div>
<div class="section level3">
<h3 id="annotation-and-enrichment-analysis">4-4 Annotation and Enrichment analysis<a class="anchor" aria-label="anchor" href="#annotation-and-enrichment-analysis"></a>
</h3>
<p>Then we select a lists of DGEs and clusters to annotate. By default
the annotation will be done on all selected DGE lists and clusters.</p>
<p><img src="Images/capture_annotation.png" style="width:80.0%"></p>
</div>
<div class="section level3">
<h3 id="annotation-and-enrichment-analysis-using-clusterprofilr">4-5 Annotation and Enrichment analysis using clusterProfilR<a class="anchor" aria-label="anchor" href="#annotation-and-enrichment-analysis-using-clusterprofilr"></a>
</h3>
<p style="text-align: justify;">
</p>
<p>We select the lists of DE genes or clusters to annotate. All
available lists are selected by default.</p>
<p style="text-align: justify;">
</p>
<p>The enrichment analysis is performed using the <a href="https://www.sciencedirect.com/science/article/pii/S2666675821000667" class="external-link">clusterProfiler
R-package</a>. For more information on the methods and the package,
please see <a href="https://www.sciencedirect.com/science/article/pii/S2666675821000667" class="external-link">The
article</a>, <a href="https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html" class="external-link">Bioconductor
Package</a> and the <a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">clusterProfiler
Vignette</a>.</p>
<p>You can chose between several domains, depending on several
parameters:</p>
<ul>
<li>custom: preferred option. A second panel of settings will ask for an
annotation file (tabulated, at least two columns, one for the name of
the gene, the other for the name or id of the annotation term) and a
third panel will allow the user to indicate which columns indicates
what. You can have as many domains/ontology as you want, just indicate
the columns name that differentiate them in the right area;</li>
<li>GO (or <a href="GO:BP" class="uri">GO:BP</a>, <a href="GO:CC" class="uri">GO:CC</a>, <a href="GO:MF" class="uri">GO:MF</a> if only one
GO domain is of interest). The user will have to chose a database
library (org.*.db) installed on the libPath and the keytype of the
entities (to make the connection with the annotation).<br>
</li>
<li>KEGG: In the second setting panel, enter the three letters to
identify the organism and the keytype of the entities. It requires to
have an online connection.</li>
</ul>
<p>Once everything is set and the analysis is conducted, several panels
are displayed to show the results.</p>
<ul>
<li>Overview: shows a table with all results for all lists analyzed, for
all domains (i.e. the number of terms found enriched). The light blue
line indicates the current enrichment.</li>
<li>Result Table: Details of the term for the current list. You can
change the domain at the bottom.</li>
<li>DotPlot: shows the first XX terms of a specified domain ordered by
adjusted p-values.</li>
<li>Heatplot: shows the first XX terms (by adjusted pvalues), the
involved genes in a heatmap. Colors are set according to the log2FC if
the list is based on a differential analysis results or left gray if it
is from coseq analysis.</li>
<li>cnetplot: shows the first XX terms (by adjusted pvalues) in beige
nodes and the involved genes in the form of a network. Colors are set
according to the log2FC if the list is based on a differential analysis
results or left gray if it is from coseq analysis.</li>
</ul>
<p style="text-align: justify;">
</p>
<p>For most of the results, you can adjust the number of terms to plot
and search for an expression to retrieve specific enriched terms.</p>
<div class="figure">
<img src="Images/Image_clusterprof1.png" style="width:80.0%" alt=""><p class="caption">Results of the analysis for multiple lists</p>
</div>
<div class="figure">
<img src="Images/Image2_clusterprof2.png" style="width:80.0%" alt=""><p class="caption">Results displayed when using a differential analysis
list</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="integration">5. Integration<a class="anchor" aria-label="anchor" href="#integration"></a>
</h2>
<p style="text-align: justify;">
Once at least two lists of DE genes are validated in the differential
analysis tabs, the Integration step will appear in the side bar menu.
Three sub menus are available.
</p>
<div class="section level3">
<h3 id="dataset-analysis-summary">5-1 Dataset analysis summary<a class="anchor" aria-label="anchor" href="#dataset-analysis-summary"></a>
</h3>
<p style="text-align: justify;">
This panel summarizes the effects of the filtering steps on the selected
datasets and the results of the differential analyses. The first row
shows how many samples and entities are left in each table. The
integration will be performed on these datasets. The second row shows
the number of DE entities for each selected contrast and for each table,
with the distinction between up-regulated and down-regulated entities.
</p>
</div>
<div class="section level3">
<h3 id="mofa">5-2 MOFA<a class="anchor" aria-label="anchor" href="#mofa"></a>
</h3>
<p style="text-align: justify;">
This integration analysis is performed using the <strong>MOFA2</strong>
R-package. It is an unsupervised method that searches for latent factors
to decompose the variance of the data, similarly to a PCA, but applied
to a multi-dataset situation. The interpretation of the results is also
very similar to those of a PCA: the most interesting factors are the one
explaining more variance, whether it’s on several dataset or just one
table. It also gives weights to the features of each table, for each
factor. This is interesting in case a factor can be <strong>a
posteriori</strong> linked to a covariate (in our case, the temperature
seems linked to the first factor, the features used to build this factor
are to be considered very relevant in the analysis).
</p>
<div class="figure">
<img src="Images/MOFA_temperaturelinked.png" style="width:80.0%" alt=""><p class="caption">MOFA results on the ecoseed datasets</p>
</div>
<p style="text-align: justify;">
MOFA allows for an unbalanced design, so it can be used even if some
samples were deleted in the previous steps.
</p>
<p style="text-align: justify;">
The user can select which tables they want to integrate, then the lists
of contrasts they want to select the DE entities from. The union or the
intersection of these lists can then be used for the analysis.
</p>
<p style="text-align: justify;">
As of now, the function used in this step only takes gaussian data. The
RNASeq data, given in the form of counts, have to be transformed to fit
the process. This is done with the <strong>voom</strong> transformation
from the <strong>limma</strong> R package.
</p>
<p style="text-align: justify;">
As recommended by MOFA FAQ, batch effects are removed before applying
the function, using <strong>removebatcheffect</strong> from
<strong>limma</strong>. See <a href="https://biofam.github.io/MOFA2/faq.html" class="external-link">MOFA-FAQ</a>. For now,
only two batch effects can be removed.
</p>
<p style="text-align: justify;">
The user can chose to scale or not each tables, with the scale views
argument, then set the number of factors to be computed and finally the
maximum iterations for the MOFA run. We recommend that the user do not
change the number of iterations.
</p>
<p style="text-align: justify;">
Once everything is set, you can run the analysis by clicking on the run
button. Several figure are provided by the interface once the analysis
is done:
</p>
<ul>
<li>Overview: gives an overview of the data, with the number of samples
and the number of entities selected for the analysis. This can be a good
step to check if the number of features in each table is in the same
range.</li>
<li>Factors Correlation: this is an important graph to check the
validity of the results. It presents the correlation between all
factors. If a correlation is greater than 0.5 (except for the diagonal),
then the results cannot be trusted. It can be because of an inadequate
normalization step. You can also check here if the number of factors you
asked for has been computed. In some cases, when the user asks for a two
great number of factors, MOFA removes the factors that explain 0% of
variance in all tables.</li>
<li>Explained variance: in this panel, two graphs are displayed
<ul>
<li>The first one shows the total explained variance for each
table.</li>
<li>The second one is the decomposition of the first for each factors.
You can see here which factors are specific to a datatable and which are
using features from all the tables.</li>
</ul>
</li>
<li>Weights plot and Weights table: can be used to check the important
features used to build each factor.</li>
<li>Factor plot: in this panel, the samples are displayed according to
their factor score. Several options are offered to the user to modify
the graph and check for any association between a factor and a
covariate.</li>
<li>Heatmap: Can only be displayed for each dataset and each factor.
Allow to represent the original data or the factor-weight combination
(denoise set to TRUE) of the most important features (weight wise) used
to built a factor. Metadata can be used to annotate the heatamp.</li>
<li>Network: not available yet!</li>
</ul>
</div>
<div class="section level3">
<h3 id="mixomics">5-3 MixOmics<a class="anchor" aria-label="anchor" href="#mixomics"></a>
</h3>
<p>Work in progress!</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nadia Bessoltane, Delphine Charif, Gwendal Cueff, Audrey Hulot, Christine Paysant-Leroux.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
