<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>RFLOMICS • RFLOMICS</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="RFLOMICS">
<meta property="og:description" content="RFLOMICS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RFLOMICS</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/RFLOMICS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/RFLOMICS-input-data.html">RFLOMICS input data</a>
    </li>
    <li>
      <a href="../articles/RFLOMICS-configure-cluster-access.html">RFLOMICS remote computing</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://forgemia.inra.fr/flomics/rflomics" class="external-link">
    <span class="fa fa-gitlab"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>RFLOMICS</h1>
            
      
      
      <div class="hidden name"><code>RFLOMICS.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p style="text-align: justify;">
      Here, we will walk through a multi-omics data analysis with the RFLOMICS package. We will see how to set up the statistical framework of the designed protocol: ie translate biological condition comparisons into contrasts, perform quality controls and statistical analysis of each omics dataset: ie. find differentially expressed genes/proteins/metabolites (DEF) per contrasts, identify clusters of co-expressed genes and perform a gene set enrichment analysis from lists of DEF or from clusters of co-expressed genes/proteins/metabolites.
      </p>
    </div>
    
<p><br><br><br></p>
<div class="section level2">
<h2 id="install-rflomics-via-docker">1. Install RFLOMICS via Docker:<a class="anchor" aria-label="anchor" href="#install-rflomics-via-docker"></a>
</h2>
<ul>
<li>install Docker</li>
</ul>
<p><a href="https://docs.docker.com/engine/install/" class="external-link uri">https://docs.docker.com/engine/install/</a></p>
<ul>
<li>Get the Dockerfile</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">https</span>://forgemia.inra.fr/flomics/rflomics/-/blob/develop.0.1/Dockerfile</span></code></pre></div>
<ul>
<li>Build the Docker Image</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> build --file=Dockerfile --tag=rflomics .</span></code></pre></div>
<ul>
<li>Run Docker</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> run -it -p 3838:3838 -v <span class="va">${HOME}</span>:/root --name=<span class="st">'rflomics'</span> --cpus 4 rflomics</span></code></pre></div>
<ul>
<li>Open a web navigator and paste this url:</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">http</span>://0.0.0.0:3838</span></code></pre></div>
<p>In the left side of interface the different tab menu will appear as the analysis progresses to guide the user.</p>
</div>
<div class="section level2">
<h2 id="load-data">2. Load Data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h2>
<p>We start by set project name (its mandatory). Then we load the experimental design and the different omics datasets.</p>
<div class="section level3">
<h3 id="dataset-example-description">2-1 Dataset example description:<a class="anchor" aria-label="anchor" href="#dataset-example-description"></a>
</h3>
<p style="text-align: justify;">
Data that will be used for this vignette have been provided by Pr. Loic Rajjou and Gwendal Cueff. They are included in the inst/ExampleFiles/ecoseed directory of the package. Briefly, A. thaliana’s transcriptom, metabolom and proteom have been obtained in the context of the study of seed germination and vigor. In particular, the author were interested by deciphering key entities involved in response to environmental stresses (on the mother plant): influence of temperature (high, medium and low) and imbibition stage (Dry: DI, early imbibition: EI and late imbibition: LI). see(<a href="https://www.uibk.ac.at/botany/ecoseed/home/" class="external-link uri">https://www.uibk.ac.at/botany/ecoseed/home/</a>)
</p>
</div>
<div class="section level3">
<h3 id="load-experimental-design">2-2 Load Experimental Design<a class="anchor" aria-label="anchor" href="#load-experimental-design"></a>
</h3>
<p>The experimental design file contains information about the readiness status of each sample (see link : vignette input ). As soon as this file is loaded, we can define the type (Biological factor : Bio; batch factor : batch; metadata information : meta) and reference level of each design factor.</p>
<ul>
<li><p><em>It is required to have minimum 1 biological factor (maximum 3 biological factors) and 1 batch factor. The order of the columns does not matter.</em></p></li>
<li><p><em>It possible to exclude samples from design. These samples will be excluded from the analysis even if they are present in the omics dataset matrix.</em></p></li>
</ul>
<div class="figure">
<img src="Images/Capture_Load-Data.jpg" style="width:80.0%" alt=""><p class="caption">interface capture</p>
</div>
</div>
<div class="section level3">
<h3 id="load-omics-data">2-3 Load Omics data<a class="anchor" aria-label="anchor" href="#load-omics-data"></a>
</h3>
<p style="text-align: justify;">
</p>
<p>For this tutorial we provide 3 datasets : RNAseq data (read counts), proteomics data (abundance of proteins), and metabolomics data (intensity of metabolites) (see link : vignette input). We must precise for each dataset its type of omics. We can give a suitable names to our dataset. By Default, they will be called respectively (set1, set2, set3).</p>
As soon as data was loaded the <strong>RFlomics MultiAssayExperiment</strong> object is created (see link : data management). The upset representation provide detail about sample intersection of different datasets.

</div>
<div class="section level3">
<h3 id="completeness-check">2-4 Completeness check<a class="anchor" aria-label="anchor" href="#completeness-check"></a>
</h3>
<p style="text-align: justify;">
</p>
<p>The design must be complete and as possible balanced. A <strong>Complete design (mandatory)</strong> mean that all possible combinations of biological factor’s level (called groups) where presents. A <strong>Balanced design (recommended)</strong> has an equal number of observations (replicates) for all groups.</p>
This constraint is checked and represented as graph (each square represent 1 group and each color level indicate the number of observations).

<p><img src="Images/completBalanced.png" width="50%"><img src="Images/completNonBalanced.png" width="50%"></p>
</div>
</div>
<div class="section level2">
<h2 id="set-the-statistical-framework">3. Set the statistical framework<a class="anchor" aria-label="anchor" href="#set-the-statistical-framework"></a>
</h2>
<p style="text-align: justify;">
</p>
<p>Interface provides models written from the simplest one to the most complete one. Only the two orders interaction terms between the biological factors will appear. Batch factor will never appear in interaction terms. In our case we choose the complete model.</p>
<p>All ‘a priori’ hypothesis (contrast) calculated from the chosen model are displayed (see link vignette contrast). The user must select those that correspond to the biological issues.</p>
<p>In our case we are studying the effect of temperature on seed germination (transition from the state of dormancy to the state of germination vigor). For that we select 3 average contrasts (see image) :</p>
This statistical framework will be applied on all loaded datasets.

<div class="figure">
<img src="Images/Capture_Experimental-Design.jpg" style="width:80.0%" alt=""><p class="caption">interface capture</p>
</div>
</div>
<div class="section level2">
<h2 id="omics-data-analysis">4. Omics data analysis<a class="anchor" aria-label="anchor" href="#omics-data-analysis"></a>
</h2>
<p>3 Item menus appear in side bar menu that corresponds to the 3 loaded omics types.</p>
<p>For each dataset, different steps of data analysis are proposed as a tab panels. This analysis steps must be performed sequentially. But we can switch between datasets. However, you cannot perform a new task until the previous one is completed. It will have a progress bar that indicates that.</p>
<div class="section level3">
<h3 id="data-exploratory">4-1 Data Exploratory<a class="anchor" aria-label="anchor" href="#data-exploratory"></a>
</h3>
<p>In this tab panel we access a data quality and we perform the appropriate data filtering and processing according to the type of omics.</p>
<p>By default, features with 0 count in all conditions are removed from the data.</p>
<p>Specifically for each data we have possibility to exclude samples or to select a subset of samples to analyze, provided the completeness condition is met.</p>
<p>The effect of data filtering or sample removing step can be explored thanks to many QC graphs. For each QC, a graph is build on each raw data and filtered data to allow the comparison.</p>
<div class="section level4">
<h4 id="rnaseq-data">RNAseq data<a class="anchor" aria-label="anchor" href="#rnaseq-data"></a>
</h4>
<p>For RNAseq data, low expressed genes are filtered such that only genes whose cpm expression is greater than <strong>“CPM cutoff”</strong> in x samples are kept. The value of x is given by <strong>“Filtering strategy”</strong> (NbOfsample_over_cpm &lt;= NbConditions). By default the data are filtered based on NbCondition strategy avec with CPM cutoff equel to 5. In this example, there are 9 conditions. So, the number of genes with a CPM less than 5 in at least 9 samples are removed.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># geneToKeep &lt;- counts[rowSums(edgeR::cpm(counts) &gt;= CPM_Cutoff) &gt;= x, ]</span></code></pre></div>
<p>To correct for differences in sequencing depths (library size), RFlomics propose one method for normalization, <strong>“TMM method”</strong> from edgeR package.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># dge &lt;- edgeR::DGEList(counts=counts, group=groups)</span>
<span class="co"># dge &lt;- edgeR::calcNormFactors(dge, method="TMM")</span></code></pre></div>
<p>For RNAses data the QC graphs are:</p>
<ul>
<li>The library size distribution which</li>
<li>Data Summary:</li>
<li>Distribution:</li>
<li>Distribution density</li>
<li>Principal component analysis (1/2)</li>
<li>Principal component analysis (2/2)</li>
</ul>
</div>
<div class="section level4">
<h4 id="proteomic-data">Proteomic data<a class="anchor" aria-label="anchor" href="#proteomic-data"></a>
</h4>
</div>
<div class="section level4">
<h4 id="metabolomic-data">Metabolomic data<a class="anchor" aria-label="anchor" href="#metabolomic-data"></a>
</h4>
</div>
</div>
<div class="section level3">
<h3 id="differential-expression-analysis">4-2 Differential expression analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis"></a>
</h3>
<p>In this tab panel, we can run the differential expression analysis. The analysis is run by contrast.<br>
For RNAseq data, the <strong>glmFit()</strong> function from the <strong>edgeR</strong> package is used whereas for proteomic and metabolomic data, the <strong>lmFit()</strong> function from <strong>limma</strong> packages is used. The ability to run analysis on a cluster is offered by sliding the cluster button (see RFLOMICS-configure-cluster-access vignette)</p>
<p>Results will appear in two components: a scrolling menu will give the results of the analysis by contrat and, at the bottom of the panel, the intersecting sets of the lists of DEG is given thanks to the <strong>UpSetR()</strong> function.</p>
<p>Each results has to bee validate:</p>
<ul>
<li>You <strong>have to</strong> to check the Pvalue distribution for each contrast and validate it (<strong>ok checkbox</strong>).
<ul>
<li>If the distribution is uniform: it is ok. a few genes are DE and the FDR will find them.</li>
<li>If the distribution as a peak near 0 and then is uniform: it is also ok. The higher is the peak, the higher is the number of DE genes. FDR correction can be applied.</li>
<li>If the distribution has 2 peaks one near 0 and one near 1 or just one peak near 1, it is not a good news ! You have to understand what’s happened.</li>
</ul>
</li>
</ul>
<div class="figure">
<embed src="Images/Pvalue_distrib.pdf" style="width:80.0%"></embed><p class="caption">Pvalue-Distrib</p>
</div>
<ul>
<li><p>You can filter diferentially expressed genes/proeins/metabolites either by fold change (<strong>|FC|</strong>) or by False Discovery Rate (BH) cut-off. Graphs will be automatically updated.</p></li>
<li><p>You have to <strong>validate</strong> to save the cut-off thresholds but also the contrast’s result selection and pass to the co-expression analysis or to the annotation enrichment step.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="coexpression-analysis">4-3 Coexpression analysis<a class="anchor" aria-label="anchor" href="#coexpression-analysis"></a>
</h3>
<ul>
<li><p>In this tab panel, choice is given to merge (<strong>union</strong>) or intersect (<strong>intersection</strong>) the list of diferentially expressed genes/proeins/metabolites associated to contrasts.</p></li>
<li>
<p>Co-expression analysis is done thanks to the <strong>coseq</strong> package with a set of optimal parameters.</p>
<ul>
<li>For <strong>RNAseq</strong>, data transformation method is set to the <strong>arcsin</strong> function and normalization method to <strong>TMM</strong>. No scale is done for this data. <strong>gaussian mixture model</strong> is used to decipher the different profils of gene expression with a set of parameters to estimate fixed to (Gaussian_pK_Lk_Ck).</li>
<li>For <strong>proteomics/metabolomics</strong> data, scaling will be done by proteins/metabolites. Another sets of parameters estimation is also proposed for the modeling of gene expression profile: (Gaussian_pK_Lk_Bk). It has to be used, if the other one doesn’t fit.</li>
</ul>
</li>
</ul>
<p>The number of technical replicates to perform (<strong>iter</strong>) for each <strong>K</strong> which is the number of expression’s profile into which the DEG have to be cut, can be first set to 2 to precise the range of K and then put to a minimum of 20 replicates per K.</p>
<p>The ability to run the clustering analysis on a cluster is offered by sliding the Cluster button (see Rflomics_configure-cluster-access).</p>
<ul>
<li>
<p>Results description (see <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/coseq/inst/doc/coseq.html" class="external-link">CoSeq</a>)</p>
<ul>
<li><p>Results can be explored with severals plots proposed by <strong>coseq</strong>. The ICL graph has been slightly modified to show all the range of the ICL values for a given K.</p></li>
<li><p>A table of jobs summary is also given. It groups error messages by K.</p></li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="annotation-and-enrichment-analysis">4-4 Annotation and Enrichment analysis<a class="anchor" aria-label="anchor" href="#annotation-and-enrichment-analysis"></a>
</h3>
<p>Then we select a lists of DGEs and clusters to annotate. By default the annotation whill be done on all setected DGE lists and clusters.</p>
</div>
</div>
<div class="section level2">
<h2 id="integration">5. Integration:<a class="anchor" aria-label="anchor" href="#integration"></a>
</h2>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nadia Bessoltane, Delphine Charif, Gwendal Cueff, Audrey Hulot, Christine Paysant-Leroux.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.4.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
