<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>abundanceBoxplot — abundanceBoxplot,SummarizedExperiment-method • RFLOMICS</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="abundanceBoxplot — abundanceBoxplot,SummarizedExperiment-method" />
<meta property="og:description" content="This function produces boxplots from raw and normalized data matrix. One color by level
of combination factor. It allows to detect outlier samples and to see the normalization
effect." />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RFLOMICS</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/RFLOMICS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/RFLOMICS-input-data.html">RFLOMICS input data</a>
    </li>
    <li>
      <a href="../articles/RFLOMICS-contrast.html">RFLOMICS contrast</a>
    </li>
    <li>
      <a href="../articles/RFLOMICS-install.html">RFLOMICS install</a>
    </li>
    <li>
      <a href="../articles/RFLOMICS-data-management.html">RFLOMICS data management</a>
    </li>
    <li>
      <a href="../articles/RFLOMICS-benchmark-coseq.html">RFLOMICS benchmark coseq</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/RFLOMICS-command-lines.html">RFLOMICS command lines</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>abundanceBoxplot</h1>
    
    <div class="hidden name"><code>abundanceBoxplot.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function produces boxplots from raw and normalized data matrix. One color by level
of combination factor. It allows to detect outlier samples and to see the normalization
effect.</p>
    </div>

    <div class="ref-usage sourceCode"><pre class='sourceCode r'><code><span class='co'># S4 method for SummarizedExperiment</span>
<span class='fu'>abundanceBoxplot</span><span class='op'>(</span><span class='va'>object</span><span class='op'>)</span></code></pre></div>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>An object of class SummarizedExperiment-class</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>#' @title Data_Distribution.plot
#'
#' @param object An object of class SummarizedExperiment
#' @param plot Type of plot : boxplot ("boxplot") or density plot ("density")
#' @exportMethod Data_Distribution.plot
#' @export
#' @return plot
#' @rdname Data_Distribution.plot
#' @examples
#' @importFrom ggplot2 geom_density boxplot xlab</p>
<p>setMethod(f="Data_Distribution.plot",
          signature="SummarizedExperiment",
          definition &lt;- function(object, plot = "boxplot")            switch (object@metadata$omicType,
                    "RNAseq" =                       # before normalization
                      if(is.null(object@metadata[["Normalization"]]$coefNorm))
                        pseudo &lt;- log2(SummarizedExperiment::assay(object) + 1) <!-- %&gt;% reshape2::melt() -->
                        y_lab  &lt;- "log2(gene counts)"
                        title  &lt;- "Raw data"                      # after normalization
                      else
                        pseudo &lt;- log2(scale(SummarizedExperiment::assay(object), center=FALSE,
                                             scale=object@metadata[["Normalization"]]$coefNorm$norm.factors)+1) <!-- %&gt;% reshape2::melt() -->
                        y_lab  &lt;- "log2(normalized gene counts)"
                        title  &lt;- "Filtered and normalized (TMM) data"                    ,
                    "proteomics" = 
                      # before rflomics transformation (plot without log2; because we don't know if input prot/meta are transformed or not)
                      if(is.null(object@metadata$transform_method))
                        pseudo &lt;- SummarizedExperiment::assay(object) <!-- %&gt;% reshape2::melt() -->
                        x_lab  &lt;- "Protein abundance (?)"
                        title  &lt;- "Raw data"                      # after transformation
                      else
                        # if log2 transformation was chosen
                        switch (object@metadata$transform_method,
                                "log2" = 
                                  pseudo &lt;- SummarizedExperiment::assay(object) <!-- %&gt;% reshape2::melt() -->
                                  x_lab  &lt;- "Transformed protein abundance" 
                                  title  &lt;- "Transformed data (method : log2)" ,                                "none" = 
                                  pseudo &lt;- log2(SummarizedExperiment::assay(object) + 1) <!-- %&gt;% reshape2::melt() -->
                                  x_lab  &lt;- "Transformed protein abundance" 
                                  title  &lt;- "Transformed data (method : ?)"  )                    ,
                    "metabolomics" = 
                      # before rflomics transformation (plot without log2; because we don't know if input prot/meta are transformed or not)
                      if(is.null(object@metadata$transform_method))
                        pseudo &lt;- SummarizedExperiment::assay(object) <!-- %&gt;% reshape2::melt() -->
                        x_lab  &lt;- "Metabolite abundance (?)"
                        title  &lt;- "Raw data"                      # after transformation
                      else
                        # if log2 transformation was chosen
                        switch( object@metadata$transform_method,
                                "log2" = 
                                  pseudo &lt;- SummarizedExperiment::assay(object) <!-- %&gt;% reshape2::melt() -->
                                  x_lab  &lt;- "Transformed protein abundance"
                                  title  &lt;- "Transformed data (method : log2)" ,
                                "none" = 
                                  pseudo &lt;- log2(SummarizedExperiment::assay(object) + 1) <!-- %&gt;% reshape2::melt() -->
                                  x_lab  &lt;- "Transformed protein abundance"
                                  title  &lt;- "Transformed data (method : ?)"  )            )            colnames(pseudo) &lt;- c("features", "samples", "value")
            pseudo.gg &lt;- dplyr::full_join(pseudo, object@metadata$Groups, by="samples")             switch (plot,
              "boxplot" = 
                #
                pseudo_bis$samples &lt;- factor(pseudo_bis$samples, levels = unique(pseudo_bis$samples))                p &lt;- ggplot(pseudo_bis, aes(x=samples, y=value)) + ggplot2::geom_boxplot(aes(fill=groups)) +
                  theme(axis.text.x = element_text(angle = 45, hjust = 1), legend.position = "none") + xlab("") + ylab(y_lab) + ggtitle(title)              ,
              "density" = 
                  # 
                  p &lt;- ggplot2::ggplot(pseudo.gg) + geom_density(aes(x=value, group = samples, color=groups), trim=FALSE) + xlab(y_lab) +
                    theme(legend.position='none') + ggtitle(title)            )            print(p)
          
)</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by First Last.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


