<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>RunDiffAnalysis — RunDiffAnalysis,SummarizedExperiment-method • RFLOMICS</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="RunDiffAnalysis — RunDiffAnalysis,SummarizedExperiment-method"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RFLOMICS</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/RFLOMICS.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/RFLOMICS-input-data.html">RFLOMICS input data</a>
    </li>
    <li>
      <a href="../articles/RFLOMICS-configure-cluster-access.html">RFLOMICS remote computing</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://forgemia.inra.fr/flomics/rflomics" class="external-link">
    <span class="fa fa-gitlab"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>RunDiffAnalysis</h1>
    
    <div class="hidden name"><code>RunDiffAnalysis-SummarizedExperiment-method.Rd</code></div>
    </div>

    <div class="ref-description">
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S4 method for SummarizedExperiment</span></span>
<span><span class="fu">RunDiffAnalysis</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">design</span>,</span>
<span>  Adj.pvalue.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  <span class="va">contrastList</span>,</span>
<span>  <span class="va">DiffAnalysisMethod</span>,</span>
<span>  clustermq <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>an object of class [SummarizedExperiment]}

{Adj.pvalue.method}{The method choosen to adjust pvalue.}

{contrastList}{The list of contrast to test}

{DiffAnalysisMethod}{A character vector giving the name of the differential analysis method
to run. Either "edgeRglmfit", "limmalmFit", ...}

{clustermq}{A boolean indicating whether the constrasts have to be computed in local or in a distant machine}

{Adj.pvalue.cutoff}{The adjusted pvalue cut-off}

{filter_only}{A boolean indicating whether only filter on DE results have to be applied (<code>filter_only=TRUE</code>). FALSE by default.}
}
{
All the results are stored as a named list <code>DiffExpAnal</code> in the metadata slot of a
given <code>SummarizedExperiment</code> object.
Objects are:</p><ul><li><p>contrasts: The selected contrasts for which the differential analysis has been conducted</p></li>
<li><p>method: The method used for the differential analysis</p></li>
<li><p>Adj.pvalue.method: The method applied for the pvalue adjustment</p></li>
<li><p>Adj.pvalue.cutoff: The threshold applied for the pvalue adjustment</p></li>
<li><p>FDR: The false discovery rate given in input</p></li>
<li><p>RawDEFres: a list giving for each contrast the raw results of the differential analysis method</p></li>
<li><p>DEF: a list giving for each contrast a data.frame of non filtered differential expressed features</p></li>
<li><p>TopDEF: a list giving for each contrast a data.frame of differential expressed features by Adj.pvalue.cutoff</p></li>
<li><p>mergeDEF: A data frame indicating for each features in row, if it is DE in a given contrasts in column</p></li>
</ul><p>An object of class [SummarizedExperiment]</p></dd>

</dl><p>This is an interface method which run a differential analysis method on
omic datasets stored in an object of class SummarizedExperiment.
According to the type of omic and to a list of contrasts,
a differential analysis method is applied to each contrasts (or hypothesis).
Three methods are available according to the type of object:
</p><ul><li><p>For RNAseq data: the <code>glmFit</code> function of the <code>edgeR</code> package</p></li>
<li><p>For proteomic and metabolomic data: the <code>lmFit</code> function of the <code>limma</code> package</p></li>
</ul>
Parameters used for RNAseq are those recommended in DiCoExpress workflow (see the paper in reference)


<p>Lambert, I., Paysant-Le Roux, C., Colella, S. et al. DiCoExpress: a tool to process multifactorial RNAseq experiments from quality controls to co-expression analysis through differential analysis based on contrasts inside GLM models. Plant Methods 16, 68 (2020).</p>


    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Nadia Bessoltane, Delphine Charif, Gwendal Cueff, Audrey Hulot, Christine Paysant-Leroux.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer></div>

  


  

  </body></html>

