<br>

## `r paste0("3.",num, " ", data)` 

```{r }

opts_chunk$set(
  fig.path=explorDir[[data]]
)
```

```{r}
dataset.raw.SE <- getRflomicsSE(rflomics.MAE, paste0(data,".raw")) 
dataset.SE     <- getRflomicsSE(rflomics.MAE, data)
variableName   <- RFLOMICS:::.omicsDic(dataset.SE)[["variableName"]]
```

* **Completness:**

`r if(length(colnames(dataset.raw.SE)) - length(colnames(dataset.SE)) == 0) {"\\begin{comment}"}`

The *outlier samples* are removed: `r setdiff(colnames(dataset.raw.SE), colnames(dataset.SE))`.

`r if(length(colnames(dataset.raw.SE)) - length(colnames(dataset.SE)) == 0) {"\\end{comment}"}`

```{r check-design, out.width = "60%", fig.align='center', fig.cap=paste0("<p style=\"color:grey\">","*fig: Number of samples per combination of factor modalities. The experimental design must be complete and ideally balanced*", "</p>")}

plotExpDesignCompleteness(dataset.SE)
```

<br>

* **Methods:**

```{r M_M_03_rna, results='asis', eval=(getOmicsTypes(dataset.SE) == "RNAseq")}

settings <- getFilterSettings(dataset.SE)

filtering <- paste0(
  "By default unexpressed genes were removed (", 
  length(metadata(dataset.raw.SE)$DataProcessing$rowSumsZero)," transcript).",
  " The low counts genes with a ", settings$method, 
  " lower to ", settings$cpmCutoff, " (CMP cutoff)",
  " in at least \"", settings$filterStrategy, "\" (filter strategy)",
  " were filtered out (",length(getFilteredFeatures(dataset.SE)),
  " transcript)."
)

Normalisation <- paste0(
  "The **TMM (Trimmed Mean of M-values)** method from **edgeR** package (version ",
  packageVersion("edgeR"), ") was applied to normalise the RNAseq libraries.")

cat("\n", filtering, " ", Normalisation, "\n")
```


```{r M_M_03_prot, results='asis',  eval=(getOmicsTypes(dataset.SE) != "RNAseq")}

Transformation <- ""
if(getTransSettings(dataset.SE)$method != "none"){
  Transformation <- paste0(
    "The data were transformed using **",
    getTransSettings(dataset.SE)$method,"**.")
}else{
  Transformation <- "The data are not transformed."
}
    
Normalisation <- ""
if(getNormSettings(dataset.SE)$method != "none"){
  Normalisation <- paste0(
    "The **", getNormSettings(dataset.SE)$method, 
    "** method was applied to normalise data.")
}else{
  Normalisation <- "The data are not normalised."
}

cat("\n", Transformation, " ", Normalisation, "\n")
```

<br>

<!-- Distribution of feature counts before/after pre-processing : boxplot -->

### {.tabset}

* **Quality control:**

#### Distribution (boxplot):

```{r boxplot, fig.show = "hold", out.width = "50%", fig.cap=paste0("<p style=\"color:grey\">","*fig: Expect aligned medians after running the pre-processing step. Samples with shifted median may be outliers.*", "</p>")}

plotDataDistribution(dataset.raw.SE, plot = "boxplot", raw = TRUE)
plotDataDistribution(dataset.SE, plot = "boxplot", raw = FALSE)

```

#### Distribution (density)

```{r density-plot, fig.show = "hold", out.width = "50%", fig.cap=paste0("<p style=\"color:grey\">","*fig: expect a gaussian density distribution after running the pre-processing step.If a second peak is observed at the beginning of the curve, this could indicate that the low, uninformative values, have not been correctly filtered. You may increase the filtering threshold.*", "</p>")}

plotDataDistribution(dataset.raw.SE, plot = "density", raw = TRUE)
plotDataDistribution(dataset.SE, plot = "density", raw = FALSE)
```

<!-- Histograms of the library size before/after pre-processing -->

`r if(getOmicsTypes(dataset.raw.SE) != "RNAseq") {"\\begin{comment}"}`

#### Library sizes:

```{r Library_size_barplot, fig.show = "hold", out.width = "50%", eval=(getOmicsTypes(dataset.raw.SE) == "RNAseq"), fig.cap=paste0("<p style=\"color:grey\">","*fig: ???*", "</p>")}

plotLibrarySize(dataset.raw.SE, raw = TRUE)
plotLibrarySize(dataset.SE)

```

```{r export,echo = FALSE}

write.table(metadata(dataset.SE)$DataProcessing$rowSumsZero,
            file = paste0(explorDir[[data]],"Table_of_non_expressed_", 
                          variableName,".txt"),
            sep ="\t",quote=FALSE,col.names=TRUE,row.names=TRUE)

write.table(getFilteredFeatures(dataset.SE),
            file = paste0(explorDir[[data]],"Table_of_low_expressed_", 
                          variableName,".txt"),
            sep="\t",quote=FALSE,col.names=TRUE,row.names=TRUE)

```

`r if(getOmicsTypes(dataset.raw.SE) != "RNAseq") {"\\end{comment}"}`

#### PCA

```{r PCA_1_2, echo=FALSE , fig.show="hold", out.width = "50%", fig.cap=paste0("<p style=\"color:grey\">","*fig: ???*", "</p>")}

plotOmicsPCA(dataset.raw.SE, raw="raw", axes=c(1, 2), groupColor="groups")
plotOmicsPCA(dataset.SE, raw="norm", axes=c(1, 2), groupColor="groups")

```

```{r PCA_1_3, echo=FALSE , fig.show="hold", out.width = "50%", fig.cap=paste0("<p style=\"color:grey\">","*fig: ???*", "</p>")}

plotOmicsPCA(dataset.raw.SE, raw="raw", axes=c(1, 3), groupColor="groups")
plotOmicsPCA(dataset.SE, raw="norm", axes=c(1, 3), groupColor="groups")
```

### {-}