## Data pre-processing: 

<br>

```{r , echo=FALSE , message=FALSE, warning=FALSE}
# create export data
explorDir <- paste0(dataDir,"/","01-Exploratory_analysis/")
dir.create(path=explorDir)

knitr::opts_chunk$set(
  fig.path=explorDir
)


```


```{r check-design, echo=FALSE , message=FALSE, warning=FALSE, echo=FALSE, out.width = "50%"}

dataset.raw.SE <- rflomics.MAE@ExperimentList[[paste0(data,".raw")]]
dataset.SE     <- rflomics.MAE@ExperimentList[[data]]
 
check <- CheckExpDesignCompleteness(rflomics.MAE, data, colnames(rflomics.MAE[[data]]))
check$plot
           
```

`r if(length(colnames(dataset.raw.SE)) - length(colnames(dataset.SE)) == 0) {"\\begin{comment}"}`

* The *outlier samples* are removed: `r setdiff(colnames(dataset.raw.SE), colnames(dataset.SE))`.

`r if(length(colnames(dataset.raw.SE)) - length(colnames(dataset.SE)) == 0) {"\\end{comment}"}`

<!-- RNAseq data cas -->

`r if(dataset.raw.SE@metadata$omicType != "RNAseq") {"\\begin{comment}"}`

Filtering and normalization strategies were adapted from DiCoExpress @Lambert2020ta

* By default, *unexpressed genes* were filtered-out (`r length(dataset.raw.SE@metadata$rowSums.zero)` genes).
* Then, *genes* with a `r getFilterSetting(dataset.SE)$method` lower to `r getFilterSetting(dataset.SE)$cpmCutoff` in at least "`r getFilterSetting(dataset.SE)$filterStrategy`" were filtered out ( `r length(getFilteredFeatures(dataset.SE))` transcripts).
* The `r getNormSetting(dataset.SE)$method` normalization method was applied. 

`r if(dataset.raw.SE@metadata$omicType != "RNAseq") {"\\end{comment}"}`

`r if(getNormSetting(dataset.SE)$method == "none") {"\\begin{comment}"}`

* The `r getNormSetting(dataset.SE)$method` normalization method was applied. 

`r if(dataset.SE@metadata$Normalization$normalized == "none") {"\\end{comment}"}`

<!-- proteomics or metabolomics data cas -->

`r if(getTransSetting(dataset.SE)$method == "none") {"\\begin{comment}"}`

* The data were transformed using `r getTransSetting(dataset.SE)$method`.

`r if(getTransSetting(dataset.SE)$method == "none") {"\\end{comment}"}`



## Quality control: {.tabset}


```{r, echo=FALSE, message=TRUE, warning=TRUE, eval=FALSE}

DT::datatable(filt.summary, options = list(colnames=FALSE, rownames = FALSE, pageLength = 10, scrollX = T, dom = 'tip'),
              class = 'cell-border stripe')

```

<br>
<!-- Histograms of the library size before/after pre-processing -->

`r if(dataset.raw.SE@metadata$omicType != "RNAseq") {"\\begin{comment}"}`

### Library sizes:

```{r Library_size_barplot, echo=FALSE, fig.show = "hold", out.width = "50%", message=FALSE, warning=FALSE, eval=(dataset.raw.SE@metadata$omicType == "RNAseq")}

Library_size_barplot.plot(dataset.raw.SE, raw = TRUE)
Library_size_barplot.plot(dataset.SE)

```

`r if(dataset.raw.SE@metadata$omicType != "RNAseq") {"\\end{comment}"}`

<!-- Distribution of feature counts before/after pre-processing : boxplot -->

### Feature counts (Boxplot):

```{r boxplot, echo=FALSE, fig.show = "hold", out.width = "50%", message=FALSE, warning=FALSE, fig.cap=""}

Data_Distribution_plot(dataset.raw.SE, plot = "boxplot", raw = TRUE)
Data_Distribution_plot(dataset.SE, plot = "boxplot", raw = FALSE)

```

### Feature counts (density)

```{r density-plot, echo=FALSE, fig.show = "hold", out.width = "50%", message=FALSE, warning=FALSE}
Data_Distribution_plot(dataset.raw.SE, plot = "density", raw = TRUE)
Data_Distribution_plot(dataset.SE, plot = "density", raw = FALSE)
```

### PCA

```{r PCA_1_2, echo=FALSE , fig.show="hold", message=FALSE, warning=FALSE, out.width = "50%"}

RFLOMICS::plotPCA(dataset.raw.SE, PCA="raw", PCs=c(1, 2), condition="groups")
RFLOMICS::plotPCA(dataset.SE, PCA="norm", PCs=c(1, 2), condition="groups")
```

```{r PCA_1_3, echo=FALSE , fig.show="hold", message=FALSE, warning=FALSE, out.width = "50%"}

RFLOMICS::plotPCA(dataset.raw.SE, PCA="raw", PCs=c(1, 3), condition="groups")
RFLOMICS::plotPCA(dataset.SE, PCA="norm", PCs=c(1, 3), condition="groups")
```



```{r export,echo = FALSE}

write.table(dataset.SE@metadata$rowSums.zero,
             file = paste0(explorDir,"Table_of_non_expressed_", RFLOMICS:::omicsDic(dataset.SE)[["variableName"]],".txt"),
             sep ="\t",quote=FALSE,col.names=TRUE,row.names=TRUE)

write.table(dataset.SE@metadata$FilteredFeatures,
             file = paste0(explorDir,"Table_of_low_expressed_", RFLOMICS:::omicsDic(dataset.SE)[["variableName"]],".txt"),
             sep="\t",quote=FALSE,col.names=TRUE,row.names=TRUE)
 
```

