## Differential expression analysis

```{r , echo=FALSE , message=FALSE, warning=FALSE}
# create export data
anaDiffDir <- paste0(dataDir,"/","02-Diff_Expression_results/")
dir.create(path=anaDiffDir)

knitr::opts_chunk$set(
  fig.path=anaDiffDir
)

```

<!-- ## Summary table of differential expression analysis -->


Differential expression analysis was carried out for each selected contrast thanks to the `r RFLOMICS::getDiffSettings(dataset.SE)$method` model.

`r if(RFLOMICS::getOmicsTypes(dataset.raw.SE) != "RNAseq") {"\\begin{comment}"}`

Differential expression analysis strategy was adapted from @Lambert2020ta

The edgeR R-package (@robinson2010a) was used to performed the analysis. 
First, the parameters of the generalized linear model were estimate thanks to the 
estimateGLMCommonDisp, estimateGLMTrendedDisp, estimateGLMTagwiseDisp and glmFit functions.
Then, the likelihood ratio test(s) (LRT) were performed for each contrast thanks to the glmLRT 
function.

`r if(RFLOMICS::getOmicsTypes(dataset.raw.SE) != "RNAseq") {"\\end{comment}"}`

`r if(RFLOMICS::getOmicsTypes(dataset.raw.SE) != "proteomics") {"\\begin{comment}"}`

Differential expression analysis strategy was adapted from @efstathiou2017a

In a first time, parameters of the linear model were estimated thanks to limma::lmFit() function of the limma package.
In a second time, contrasts.fit was used and empirical bayes statistics were obtained thanks to the eBayes() functions of the limma r-package.
`r if(RFLOMICS::getOmicsTypes(dataset.raw.SE) != "proteomics") {"\\end{comment}"}`

`r if(RFLOMICS::getOmicsTypes(dataset.raw.SE) != "metabolomics") {"\\begin{comment}"}`
In a first time, parameters of the linear model were estimated thanks to limma::lmFit() function of the limma package.
In a second time, contrasts.fit() was used and empirical bayes statistics were obtained thanks to the eBayes() functions of the limma r-package.
`r if(RFLOMICS::getOmicsTypes(dataset.raw.SE) != "metabolomics") {"\\end{comment}"}`


P-values of differential expressed `r RFLOMICS:::omicsDic(dataset.raw.SE)[["variableName"]]` were adjusted according 
to the `r RFLOMICS::getDiffSettings(dataset.SE)$p.adj.method` procedure with a cutoff set to `r RFLOMICS::getDiffSettings(dataset.SE)$p.adj.cutoff`.
LogFC cut-off was set to `r RFLOMICS::getDiffSettings(dataset.SE)$abs.logFC.cutoff`


* **Summary** 

```{r , out.width = "100%", fig.show = "hold", eval=Diff.eval}

tmpRes <- dataset.SE@metadata[["DiffExpAnal"]]
DE <- vector()
DEup <- vector()
DEdown <- vector()

for (i  in  1:dim(tmpRes[["contrasts"]])[1]){
  contrastName <- tmpRes[["contrasts"]][i,]$contrastName
  stats <- tmpRes[["stats"]][contrastName,]
  DE[i] <- stats[["All"]]
  DEup[i] <- paste(stats[["Up"]]," (",round(stats[["Up"]]/stats[["All"]],2)*100," %)",sep="")
  DEdown[i] <- paste(stats[["Down"]]," (",round(stats[["Down"]]/stats[["All"]],2)*100," %)",sep="")
}

```


```{r , echo=FALSE, results="asis", eval=Diff.eval}
kableExtra::kable_styling(knitr::kable(dplyr::bind_cols(tmpRes[["contrasts"]][,c("tag","contrastName")], data.frame("DE.Tot"=DE,"DE.Up"=DEup,"DE.Down"=DEdown)), row.names=FALSE, caption=NULL, format="html"), font_size=12)
```


This **UpSetR** @Conway2017a plot gives the number of common and specific differential expressed `r RFLOMICS:::omicsDic(dataset.SE)[["variableName"]]` per contrasts.

```{r Upset_plot,echo=FALSE, message=FALSE, warning=FALSE, fig.show = "hold", out.width = "80%", echo=FALSE, eval=Diff.eval}

DEF_mat <- dataset.SE@metadata$DiffExpAnal[["mergeDEF"]]

if(dim(DEF_mat)[2] > 2){
  UpSetR::upset(DEF_mat, sets = (names(DEF_mat[,-1]))) 
}

write.table(DEF_mat, file=paste0(anaDiffDir,"/","Matrice_of_DE_", RFLOMICS:::omicsDic(dataset.SE)[["variableName"]],"_per_Constrast.txt"), sep="\t",quote=FALSE,col.names=TRUE,row.names=FALSE)


```

* **Display results by contrast(s):** 

*Please click on the arrows to scroll down the results and click on the figure to zoom*

```{r , message=FALSE, warning=FALSE, echo=FALSE, eval=Diff.eval}

out <- NULL

for (contrast  in  metadata(dataset.SE)[["DiffExpAnal"]][["contrasts"]]$contrastName) {
  out <- c(out, knitr::knit_child(paste0(path.package("RFLOMICS"),"/RFLOMICSapp/","report_04_diff_analysis_per_contrast.Rmd"), quiet=TRUE))  
  #}
}
knitr::asis_output(out)
```


